{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "unexpected indent (<ipython-input-3-5b08cc71c2f6>, line 6)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-3-5b08cc71c2f6>\"\u001b[0;36m, line \u001b[0;32m6\u001b[0m\n\u001b[0;31m    longest, x_longest = 0, 0\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m unexpected indent\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "def findHomology(sequence1, sequence2): #Returns largest homologous region between the 2 sequences\n",
    "    m = [[0] * (1+len(sequence2))\n",
    "    for i in range(1+len(sequence1))]\n",
    "        longest, x_longest = 0, 0\n",
    "        for x in range(1, 1 + len(sequence2)):\n",
    "            for y in range(1, 1 + len(sequence2)):\n",
    "                if sequence1[x - 1] == sequence2[y - 1]:\n",
    "                    m[x][y] = m[x - 1][y - 1] + 1\n",
    "                    if m[x][y] > longest:\n",
    "                        longest = m[x][y]\n",
    "                        x_longest = x\n",
    "                else:\n",
    "                    m[x][y] = 0\n",
    "    return sequence1[x_longest - longest: x_longest]\n",
    "        \n",
    "def choosePrimers(insert, upstreamBackbone, downstreamBackbone): #Make primers for an insert to amplify it such that it can fit in a \n",
    "    forwardPrimer = insert[:20]\n",
    "    reversePrimer = reverseComplement(insert[len(insert) - 20: len(insert)])\n",
    "    getTm\n",
    "    \n",
    "def reverseComplement(sequence):\n",
    "    reverseComplement = \"\"\n",
    "    for base in sequence:\n",
    "        if base == \"a\":\n",
    "            reverseComplement = reverseComplement + \"t\"\n",
    "        elif base == \"t\":\n",
    "            reverseComplement = reverseComplement + \"a\"\n",
    "        elif base == \"g\":\n",
    "            reverseComplement = reverseComplement + \"c\"\n",
    "        elif base == \"c\":\n",
    "            reverseComplement = reverseComplement + \"g\"\n",
    "        else:\n",
    "            print (\"error\")\n",
    "    return (reverseComplement)\n",
    "def getTm(sequence):\n",
    "    composition = getComposititon(sequence)\n",
    "    #Tm= 64.9 +41*(yG+zC-16.4)/(wA+xT+yG+zC) \n",
    "    Tm = 64.9 + 41*(composition[\"G\"] + composition[\"C\"] - 16.4) / (composition[\"A\"] + composition[\"T\"] + composition[\"G\"] + composition[\"C\"])\n",
    "    return (Tm)\n",
    "def getComposition(sequence):\n",
    "    composition = {\"A\" : 0, \"G\" : 0, \"C\" : 0, \"T\" : 0}\n",
    "    for base in sequence:\n",
    "        if base == \"a\":\n",
    "            composition[\"A\"] = composition[\"A\"] + 1\n",
    "        elif base == \"t\":\n",
    "            composition[\"T\"] = composition[\"T\"] + 1\n",
    "        elif base == \"g\":\n",
    "            composition[\"G\"] = composition[\"G\"] + 1\n",
    "        elif base == \"c\":\n",
    "            composition[\"C\"] = composition[\"C\"] + 1\n",
    "        else:\n",
    "            print(\"error 2\")\n",
    "    return composititon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGGAAAAGACATATAACCCACAAGATATCGAACA forward\n",
      "CAGCGCGGCGATAACAGCC reverse\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "('CTAGAAATAATTTTGTTTAACTTTAAGAAGGAGATATACCATGGAAAAGACATATAACCCACAAGATATCGAACA',\n",
       " 'CTTTGTTAGCAGCCGGATCTCAGTGGTGGTGGTGGTGGTGCAGCGCGGCGATAACAGCC')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import primer3\n",
    "\n",
    "class DNASequence():\n",
    "    def __init__(self, sequence):\n",
    "        #Initialize the DNA variable, which has sequence info as well as additional properties (Tm, homodimer & hairpin formation)\n",
    "        self.sequence = sequence\n",
    "        self.composition = self.getComposition(sequence)\n",
    "        if len(sequence) != 0:\n",
    "            self.gcContent = self.getGCcontent(sequence)\n",
    "        else:\n",
    "            self.gcContent = 0\n",
    "        if len(sequence) < 60:\n",
    "            self.Tm = primer3.calcTm(sequence)\n",
    "            self.Homodimer = primer3.calcHomodimer(sequence)\n",
    "            self.Hairpin = primer3.calcHairpin(sequence)\n",
    "        else:\n",
    "            self.Tm = 100000 #set the Tm arbitrarily high\n",
    "        \n",
    "    def getComposition(self, sequence):\n",
    "        composition = {\"A\" : 0, \"G\" : 0, \"C\" : 0, \"T\" : 0}\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                composition[\"A\"] = composition[\"A\"] + 1\n",
    "            elif base == \"T\":\n",
    "                composition[\"T\"] = composition[\"T\"] + 1\n",
    "            elif base == \"G\":\n",
    "                composition[\"G\"] = composition[\"G\"] + 1\n",
    "            elif base == \"C\":\n",
    "                composition[\"C\"] = composition[\"C\"] + 1\n",
    "            else:\n",
    "                print(\"error 2\")\n",
    "        return composition\n",
    "    def getGCcontent(self, sequence):\n",
    "        GCpercent = (self.composition[\"G\"] + self.composition[\"C\"])/(self.composition[\"G\"] + self.composition[\"T\"] + self.composition[\"A\"] + self.composition[\"C\"])\n",
    "        return GCpercent\n",
    "    def reverseComplement(self, sequence):\n",
    "        #Takes the complement of the sequence and reverses it.\n",
    "        reverseComplement = \"\"\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                reverseComplement = \"T\" + reverseComplement\n",
    "            elif base == \"T\":\n",
    "                reverseComplement = \"A\" + reverseComplement\n",
    "            elif base == \"G\":\n",
    "                reverseComplement = \"C\" + reverseComplement\n",
    "            elif base == \"C\":\n",
    "                reverseComplement = \"G\" + reverseComplement \n",
    "            else:\n",
    "                print (\"error\")\n",
    "        return (reverseComplement)\n",
    "def reverseComplement(sequence):\n",
    "    #Takes the complement of the sequence and reverses it.\n",
    "    reverseComplement = \"\"\n",
    "    for base in sequence:\n",
    "        if base == \"A\":\n",
    "            reverseComplement = \"T\" + reverseComplement \n",
    "        elif base == \"T\":\n",
    "            reverseComplement = \"A\" + reverseComplement\n",
    "        elif base == \"G\":\n",
    "            reverseComplement = \"C\" + reverseComplement\n",
    "        elif base == \"C\":\n",
    "            reverseComplement = \"G\" + reverseComplement\n",
    "        else:\n",
    "            print (\"error\")\n",
    "    return (reverseComplement)\n",
    "def screenPrimers(insert, upstreamBackbone, downstreamBackbone):\n",
    "    #run the main program to create and screen the forward and reverse primers and then add overhangs.\n",
    "    forwardPrimer = screenPrimer(insert, \"F\")\n",
    "    print(forwardPrimer.sequence, \"forward\")\n",
    "    reversePrimer = screenPrimer(insert, \"R\")\n",
    "    print(reversePrimer.sequence, \"reverse\")\n",
    "    \n",
    "    primers = addOverhangs(forwardPrimer, reversePrimer, \"reverse\", upstreamBackbone, downstreamBackbone)\n",
    "    return (primers[0].sequence, primers[1].sequence)\n",
    "def screenPrimer(insert, whichPrimer, primerSize=18): #Make primers for an insert to amplify it such that it can be assembled with Gibson\n",
    "    if primerSize < 40:\n",
    "        #Make sure you are selecting the right region to prime. \n",
    "        #This algorithm creates primers that amplify the WHOLE SEQUENCE minus the stop codon at the end.\n",
    "        if whichPrimer == \"F\": \n",
    "            primer = DNASequence(insert.sequence[:primerSize])\n",
    "        elif whichPrimer == \"R\":\n",
    "            primer = DNASequence(reverseComplement(insert.sequence[len(insert.sequence) - primerSize: len(insert.sequence)]))\n",
    "        else:\n",
    "            print(\"error\")\n",
    "        #Add various screens below here. You can update the parameters to be more or less stringent depending on how likely you would like your primers to work\n",
    "        isGood = True\n",
    "        if primer.Tm < 58.0 or primer.Tm > 59.5:\n",
    "            isGood = False\n",
    "        if float(primer.Homodimer.dg) < -10000.0:\n",
    "            isGood = False\n",
    "        if float(primer.Hairpin.dg) < -5000.0:\n",
    "            isGood = False\n",
    "    \n",
    "        #return the shortest (18bp ≤) desireable primers\n",
    "        if isGood == True:\n",
    "            #screening returned a good primer; return the primer\n",
    "            return primer\n",
    "        else:\n",
    "            #primer failed screening, try again with a larger primer using recursion\n",
    "            if whichPrimer == \"F\":\n",
    "                primer = screenPrimer(insert, \"F\", primerSize + 1) \n",
    "            elif whichPrimer == \"R\":\n",
    "                primer = screenPrimer(insert, \"R\", primerSize + 1)\n",
    "            return primer\n",
    "    else:\n",
    "        #parameters are too stringent; no primers were found that matched the criteria given. Return an empty string\n",
    "        print (\"You cannot make primers with the given parameters\")\n",
    "        return DNASequence(\"\")\n",
    "def addOverhangs(primer1, primer2, orientation, leftBackbone, rightBackbone):\n",
    "    #decide how the construct should be oriented in the final vector and add the overhangs so Gibson will add correctly.\n",
    "    if orientation == \"forward\":\n",
    "        primer2.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer2.sequence\n",
    "        primer1.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer1.sequence\n",
    "        return primer1, primer2\n",
    "    elif orientation == \"reverse\":\n",
    "        primer1.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer1.sequence\n",
    "        primer2.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer2.sequence\n",
    "        return primer1, primer2\n",
    "    else:\n",
    "        print (\"please enter either a forward or reverse strand orientation\")\n",
    "\n",
    "\n",
    "        \n",
    "insert = DNASequence('atgGAAAAGACATATAACCCACAAGATATCGAACAGCCGCTTTACGAGCACTGGGAAAAGCTGATTGAACAGCAGGCTGTTATCGCCGCGCTG'.upper())\n",
    "leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg'.upper())\n",
    "\n",
    "rightbackbone = DNASequence('ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcgcgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgcca'.upper())\n",
    "\n",
    "screenPrimers(insert, leftbackbone , rightbackbone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGCGACGCATTGAAACGGTCTTCAACGCCAGAAA'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reverseComplement('TTTCTGGCGTTGAAGACCGTTTCAATGCGTCGCT'.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import primer3\n",
    "\n",
    "primer3.calcTm('atggaaaagacatataacccacaagatatcgaac'.upper())\n",
    "\n",
    "t = (reverseComplement('ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgc'[:40].upper()))\n",
    "\n",
    "print(t)\n",
    "\n",
    "leftbackbone = ('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg'.upper())\n",
    "\n",
    "def reverse(sequence):\n",
    "    reverse = \"\"\n",
    "    for base in sequence:\n",
    "        if base == \"A\":\n",
    "            reverse = \"A\" + reverse \n",
    "        elif base == \"T\":\n",
    "            reverse = \"T\" + reverse\n",
    "        elif base == \"G\":\n",
    "            reverse = \"G\" + reverse\n",
    "        elif base == \"C\":\n",
    "            reverse = \"C\" + reverse\n",
    "        else:\n",
    "            print (\"error\")\n",
    "    return (reverse)\n",
    "\n",
    "s = (leftbackbone)\n",
    "\n",
    "s[len(s) - 40: len(s)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def long_substr(data):\n",
    "    substr = ''\n",
    "    if len(data) > 1 and len(data[0]) > 0:\n",
    "        for i in range(len(data[0])):\n",
    "            for j in range(len(data[0])-i+1):\n",
    "                if j > len(substr) and is_substr(data[0][i:i+j], data):\n",
    "                    substr = data[0][i:i+j]\n",
    "    return substr\n",
    "\n",
    "def is_substr(find, data):\n",
    "    if len(data) < 1 and len(find) < 1:\n",
    "        return False\n",
    "    for i in range(len(data)):\n",
    "        if find not in data[i]:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "\n",
    "print (long_substr(['CCCCATCGATAACCCAGGCATCAAATAAAACGAAAGGCTCAGTCGAAAGACTGGGCCTTTCGT', 'CTGTAAACCGAGGTTTTGGACTTAGCACCTCTGTTTGGGAAAATTGGGGAGTCGAGTCGGATTTGATCTGATCAAGAGACAGGATGAGGATCGTTTCGCAAAGAGGAGAAAGAGCTCATGATTGAACAAGATGGATTGCACGCAGGTTCTCCGGCCGCTTGGGTGGAGAGGCTATTCGGCTATGACTGGGCACAACAGACAATCGGCTGCTCTGATGCCGCCGTGTTCCGGCTGTCAGCGCAGGGGCGCCCGGTTCTTTTTGTCAAGACCGACCTGTCCGGTGCCCTGAATGAACTTCAGGACGAGGCAGCGCGGCTATCGTGGCTGGCCACGACGGGCGTTCCTTGCGCAGCTGTGCTCGACGTTGTCACTGAAGCGGGAAGGGACTGGCTGCTATTGGGCGAAGTGCCGGGGCAGGATCTCCTGTCATCTCACCTTGCTCCTGCCGAGAAAGTATCCATCATGGCTGATGCAATGCGGCGGCTGCATACGCTTGATCCGGCTACCTGCCCATTCGACCACCAAGCGAAACATCGCATCGAGCGAGCACGTACTCGGATGGAAGCCGGTCTTGTCGATCAGGATGATCTGGACGAAGAGCATCAGGGGCTCGCGCCAGCCGAACTGTTCGCCAGGCTCAAGGCGCGCATGCCCGACGGCGAGGATCTCGTCGTGACCCACGGCGATGCCTGCTTGCCGAATATCATGGTGGAAAATGGCCGCTTTTCTGGATTCATCGACTGTGGCCGGCTGGGTGTGGCGGACCGCTATCAGGACATAGCGTTGGCTACCCGTGATATTGCTGAAGAGCTTGGCGGCGAATGGGCTGACCGCTTCCTCGTGCTTTACGGTATCGCCGCTCCCGATTCGCAGCGCATCGCCTTCTATCGCCTTCTTGACGAGTTCTTCTGAGCGGGACTCTGGGGTTCGAAATGACCGACCAAGCGACGCCCAACCTGCCATCACGAGATTTCGATTCCACCGCCGCCTTCTATGAAAGGTTGGGCTTCGGAATCGTTTTCCGGGACGCCGGCTGGATGATCCTCCAGCGCGGGGATCTCATGCTGGAGTTCTTCGCCCACCCCATCGATAACCCAGGCATCA']))\n",
    "\n",
    "#findHomology('TCCCTATCAGTGATAGAGAAAAGAGAATCGGATCCAAAAGATCTAAAGAGGAGAAAGGATCTATGTTAATTGGCTATGTCCGTGTGTCCACCAACGATCAAAACACAGATCTTCAAAGAAATGCCCTCGTTTGCGCGGGATGTGAGCAGATCTTTGAAGATAAACTTAGCGGTACCCGTACCGATCGACCGGGTCTTAAACGCGCACTTAAACGGCTTCAGAAGGGGGATACCTTGGTCGTATGGAAACTGGACCGGCTGGGTAGATCGATGAAACATCTGATTAGTTTAGTTGGAGAACTGCGAGAACGCGGCATCAATTTTCGCTCCTTAACCGATAGCATCGATACCTCCAGTCCAATGGGGAGGTTTTTCTTCTATGTTATGGGCGCGCTGGCCGAAATGGAACGTGAATTGATTATCGAGCGGACAATGGCTGGGTTAGCGGCAGCACGAAATAAAGGACGGCGCTTCGGTCGTCCACCCAAGGGCGGCTCCGGTGGCTCGGGAGGTTCCGGCGGTAGCGGGGGCAGCGGCGGCAGTGGGGGCTCAGGAGGTTCTGATAAAAAATATTCTATAGGACTGGCTATTGGTACCAATTCAGTGGGTTGGGCGGTAATTACCGACGAATATAAAGTGCCGTCTAAAAAATTTAAGGTCTTGGGGAATACAGACCGTCACAGTATAAAAAAAAATCTGATCGGTGCGCTGCTGTTTGACTCTGGAGAAACTGCCGAAGCAACTCGTTTAAAACGAACCGCACGCCGTCGCTATACAAGGCGGAAAAATCGTATATGCTATTTACAAGAAATTTTTTCCAATGAAATGGCAAAAGTCGATGACAGCTTCTTTCATCGCTTAGAAGAGTCCTTTTTAGTTGAAGAAGATAAAAAACACGAACGCCATCCAATTTTTGGGAATATAGTTGATGAAGTCGCCTATCACGAAAAATATCCGACTATTTACCACCTGCGTAAGAAACTTGTGGATTCGACGGATAAAGCAGATCTGAGGCTGATTTATCTCGCCCTCGCACATATGATTAAATTCAGAGGTCATTTCCTGATAGAAGGCGATCTGAATCCCGACAACTCTGATGTTGATAAGTTGTTTATACAGCTGGTGCAAACCTATAATCAACTATTCGAAGAGAATCCTATAAACGCCAGCGGAGTGGACGCTAAAGCGATACTCAGCGCGCGCTTATCTAAAAGCCGGCGGCTGGAAAATCTGATCGCACAATTGCCAGGTGAAAAAAAAAATGGTTTATTTGGCAATCTGATAGCGCTTTCACTGGGGCTAACACCCAATTTTAAATCGAATTTTGATTTAGCCGAAGATGCTAAGCTGCAGCTGAGTAAAGATACCTATGATGACGATCTTGATAACCTGCTGGCACAAATTGGAGATCAATACGCAGATCTGTTTTTGGCCGCCAAGAATCTGTCGGATGCAATTCTGCTGAGCGATATTTTGAGGGTGAATACTGAAATCACTAAGGCACCGCTCTCTGCAAGTATGATCAAGCGTTATGATGAGCATCACCAGGACTTAACCCTGTTAAAAGCTCTGGTCCGCCAGCAGCTCCCGGAAAAATACAAGGAAATATTCTTCGATCAATCTAAGAATGGGTATGCCGGCTATATCGATGGCGGAGCGTCTCAGGAAGAATTTTATAAATTCATAAAACCTATCCTTGAAAAAATGGATGGAACGGAAGAGCTGCTGGTTAAACTGAATAGAGAGGATCTGTTACGCAAGCAGAGGACATTCGATAACGGTAGTATCCCACATCAAATACATCTGGGTGAACTTCACGCGATTCTTCGCCGCCAGGAAGACTTTTATCCCTTTCTAAAAGACAACAGGGAGAAAATTGAAAAAATCCTGACGTTTCGTATTCCTTATTATGTGGGGCCGCTGGCTCGTGGAAACAGTCGGTTCGCATGGATGACGCGCAAGTCCGAAGAGACGATAACGCCGTGGAATTTCGAAGAAGTGGTCGACAAAGGCGCAAGTGCACAGTCATTCATCGAACGTATGACTAACTTCGATAAAAATTTACCAAACGAGAAAGTTCTGCCTAAGCATAGCCTTCTCTATGAATACTTTACAGTCTATAACGAGTTAACTAAAGTTAAATATGTGACTGAAGGTATGCGTAAACCGGCGTTTCTGAGCGGCGAACAAAAAAAGGCAATTGTCGATCTGTTATTTAAAACAAATCGTAAAGTTACGGTCAAGCAATTGAAAGAGGACTACTTCAAAAAAATTGAGTGTTTCGACAGCGTCGAAATTTCTGGCGTTGAAGACCGTTTCAAT', 'CGTTGAAGACCGTTTCAATGCGTCGCTGGGAACTTACCACGATCTACTCAAAATCATAAAAGATAAAGACTTTCTGGACAATGAAGAGAACGAAGACATACTTGAAGACATCGTTCTGACCCTTACGCTATTTGAAGATAGGGAGATGATCGAAGAGCGCTTGAAGACCTATGCGCACCTATTCGATGATAAAGTTATGAAACAGCTCAAGCGGCGCCGTTACACGGGGTGGGGGCGTCTGTCGAGGAAGTTAATCAATGGAATTCGTGATAAGCAGTCAGGCAAGACGATCCTGGACTTTCTGAAATCAGATGGGTTTGCAAATCGTAACTTCATGCAGCTGATCCATGATGATTCGTTAACTTTTAAAGAAGATATTCAAAAAGCCCAAGTATCTGGTCAGGGTGACTCTCTTCACGAACATATCGCGAATTTGGCAGGTTCACCAGCTATTAAAAAAGGTATCTTGCAGACTGTGAAAGTTGTCGATGAACTTGTTAAAGTTATGGGTCGTCACAAGCCAGAGAATATCGTAATTGAGATGGCGCGTGAAAATCAAACAACCCAAAAAGGCCAGAAAAACTCGAGGGAACGCATGAAACGTATTGAAGAAGGTATTAAGGAGTTGGGTTCCCAGATACTTAAAGAACACCCGGTCGAAAATACGCAACTGCAAAATGAAAAATTATATCTATATTATCTTCAGAACGGGCGCGATATGTATGTGGACCAAGAGCTGGATATCAATCGGCTTTCGGATTATGACGTCGATGCAATCGTTCCGCAAAGTTTCCTCAAGGATGACAGCATCGATAACAAAGTGTTAACCCGTAGCGATAAAAACCGTGGCAAAAGTGATAACGTGCCATCGGAAGAGGTAGTGAAAAAGATGAAAAATTATTGGCGCCAGCTGCTTAATGCTAAGCTGATTACGCAACGGAAATTTGATAATCTGACTAAAGCTGAGCGGGGCGGGCTGTCGGAACTCGATAAAGCTGGGTTCATAAAAAGACAATTGGTCGAGACGCGCCAGATTACCAAACACGTAGCCCAAATTCTGGATTCACGGATGAATACGAAATACGACGAGAATGATAAACTGATCCGGGAAGTGAAAGTGATCACTCTTAAATCTAAACTGGTATCAGACTTTAGAAAAGATTTCCAATTCTATAAGGTTCGTGAAATCAACAATTATCATCACGCTCATGACGCCTATTTAAATGCGGTGGTAGGCACAGCGTTAATCAAGAAATATCCGAAACTTGAGTCAGAGTTTGTGTATGGAGATTATAAAGTTTATGATGTTAGAAAAATGATTGCTAAATCGGAGCAAGAAATTGGTAAGGCTACAGCGAAATATTTTTTTTACTCCAATATTATGAATTTCTTTAAGACTGAAATTACCTTGGCTAATGGTGAAATCCGCAAACGGCCGCTGATCGAAACAAATGGTGAAACTGGCGAGATTGTTTGGGATAAGGGCCGCGATTTCGCAACGGTCCGTAAGGTGTTGAGTATGCCTCAGGTGAATATTGTAAAGAAAACGGAAGTACAGACTGGCGGGTTTAGCAAAGAATCAATTTTGCCGAAACGTAATTCTGATAAACTGATCGCTCGTAAGAAAGATTGGGATCCCAAGAAGTATGGTGGCTTTGATAGTCCGACCGTAGCGTATAGTGTTCTGGTGGTAGCTAAGGTAGAAAAAGGCAAATCAAAGAAACTGAAGTCCGTGAAGGAACTACTGGGTATCACGATCATGGAACGATCAAGTTTTGAAAAAAATCCGATCGACTTCCTTGAAGCGAAAGGGTACAAAGAGGTAAAGAAGGACCTAATTATAAAACTGCCCAAATATAGTCTGTTTGAACTGGAAAATGGAAGGAAACGAATGTTGGCAAGCGCTGGTGAGCTGCAGAAAGGCAATGAATTAGCCCTGCCCTCGAAATATGTAAATTTCTTATACTTGGCCAGTCATTATGAAAAGCTGAAAGGATCACCTGAAGATAATGAACAAAAACAGCTCTTCGTCGAGCAACACAAGCATTACTTAGATGAAATTATAGAACAAATTTCTGAATTCTCAAAACGTGTTATATTAGCGGACGCGAATCTGGATAAGGTCTTGAGCGCATATAACAAGCATCGCGATAAACCGATACGGGAACAAGCGGAAAATATTATACATCTGTTCACGCTGACCAATTTAGGCGCGCCTGCTGCGTTTAAATATTTTGATACAACCATAGATCGTAAACGTTACACCAGCACAAAGGAAGTCCTGGATGCCACACTGATCCATCAGAGCATAACAGGACTGTACGAGACCCGTATCGATCTGAGTCAGCTTGGAGGTGATGGTGGTTCGGATTACAAGGATGACGATGATAAAGGTGGAAGTCCGAAAAAAAAAAGGAAAGTTTGATAACTCGAGTAAGGATCAGTAAGGATCTCCAGGCATC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [400]>\n",
      "<Response [400]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n"
     ]
    }
   ],
   "source": [
    "import primer3\n",
    "import pandas\n",
    "import math\n",
    "import requests\n",
    "#import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "sequences = []\n",
    "breakdown = []\n",
    "\n",
    "seqPrimers = []\n",
    "seqPrimerNames = []\n",
    "\n",
    "data = pandas.read_excel(\"/Users/aleclourenco/Desktop/REU-2017-Endy-Lab/PURE Datasheet.xlsx\", sheetname=\"List of PURE Components\", skiprows=[0])\n",
    "\n",
    "class DNASequence():\n",
    "    def __init__(self, sequence):\n",
    "        #Initialize the DNA variable, which has sequence info as well as additional properties (Tm, homodimer & hairpin formation)\n",
    "        self.sequence = sequence\n",
    "        self.composition = self.getComposition(sequence)\n",
    "        if len(sequence) != 0:\n",
    "            self.gcContent = self.getGCcontent(sequence)\n",
    "        else:\n",
    "            self.gcContent = 0\n",
    "        if len(sequence) < 60:\n",
    "            self.Tm = primer3.calcTm(sequence)\n",
    "            self.Homodimer = primer3.calcHomodimer(sequence)\n",
    "            self.Hairpin = primer3.calcHairpin(sequence)\n",
    "        else:\n",
    "            self.Tm = 100000 #set the Tm arbitrarily high\n",
    "        \n",
    "    def getComposition(self, sequence):\n",
    "        composition = {\"A\" : 0, \"G\" : 0, \"C\" : 0, \"T\" : 0}\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                composition[\"A\"] = composition[\"A\"] + 1\n",
    "            elif base == \"T\":\n",
    "                composition[\"T\"] = composition[\"T\"] + 1\n",
    "            elif base == \"G\":\n",
    "                composition[\"G\"] = composition[\"G\"] + 1\n",
    "            elif base == \"C\":\n",
    "                composition[\"C\"] = composition[\"C\"] + 1\n",
    "            else:\n",
    "                print(\"error 2\")\n",
    "        return composition\n",
    "    def getGCcontent(self, sequence):\n",
    "        GCpercent = (self.composition[\"G\"] + self.composition[\"C\"])/(self.composition[\"G\"] + self.composition[\"T\"] + self.composition[\"A\"] + self.composition[\"C\"])\n",
    "        return GCpercent\n",
    "    def reverseComplement(self, sequence):\n",
    "        #Takes the complement of the sequence and reverses it.\n",
    "        reverseComplement = \"\"\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                reverseComplement = \"T\" + reverseComplement\n",
    "            elif base == \"T\":\n",
    "                reverseComplement = \"A\" + reverseComplement\n",
    "            elif base == \"G\":\n",
    "                reverseComplement = \"C\" + reverseComplement\n",
    "            elif base == \"C\":\n",
    "                reverseComplement = \"G\" + reverseComplement \n",
    "            else:\n",
    "                print (\"error\")\n",
    "        return (reverseComplement)\n",
    "def reverseComplement(sequence):\n",
    "    #Takes the complement of the sequence and reverses it.\n",
    "    reverseComplement = \"\"\n",
    "    for base in sequence:\n",
    "        if base == \"A\":\n",
    "            reverseComplement = \"T\" + reverseComplement \n",
    "        elif base == \"T\":\n",
    "            reverseComplement = \"A\" + reverseComplement\n",
    "        elif base == \"G\":\n",
    "            reverseComplement = \"C\" + reverseComplement\n",
    "        elif base == \"C\":\n",
    "            reverseComplement = \"G\" + reverseComplement\n",
    "        else:\n",
    "            print (\"error\")\n",
    "    return (reverseComplement)\n",
    "def screenPrimers(insert, upstreamBackbone, downstreamBackbone):\n",
    "    #run the main program to create and screen the forward and reverse primers and then add overhangs.\n",
    "    forwardPrimer = screenPrimer(insert, \"F\")\n",
    "    #print(forwardPrimer.sequence, \"forward\")\n",
    "    reversePrimer = screenPrimer(insert, \"R\")\n",
    "    #print(reversePrimer.sequence, \"reverse\")\n",
    "    if len(forwardPrimer.sequence) and len(reversePrimer.sequence) != 0:\n",
    "        primers = addOverhangs(forwardPrimer, reversePrimer, \"reverse\", upstreamBackbone, downstreamBackbone)\n",
    "        return (primers[0].sequence, primers[1].sequence)\n",
    "    else:\n",
    "        return (\"You cannot make primers with the given parameters\", \"You cannot make primers with the given parameters\")\n",
    "    \n",
    "def screenPrimer(insert, whichPrimer, primerSize=18): #Make primers for an insert to amplify it such that it can be assembled with Gibson\n",
    "    if primerSize < 40:\n",
    "        #Make sure you are selecting the right region to prime. \n",
    "        #This algorithm creates primers that amplify the WHOLE SEQUENCE minus the stop codon at the end.\n",
    "        noncanonicalStartCodon = False\n",
    "        if whichPrimer == \"F\": \n",
    "            primer = DNASequence(insert.sequence[:primerSize])\n",
    "            if (primer.sequence[:3].lower() != \"atg\"):\n",
    "                primer = DNASequence(insert.sequence[3:primerSize])\n",
    "                noncanonicalStartCodon = True\n",
    "        elif whichPrimer == \"R\":\n",
    "            primer = DNASequence(reverseComplement(insert.sequence[len(insert.sequence) - primerSize: len(insert.sequence)]))\n",
    "        else:\n",
    "            print(\"error\")\n",
    "        #Add various screens below here. You can update the parameters to be more or less stringent depending on how likely you would like your primers to work\n",
    "        isGood = True\n",
    "        if primer.Tm < 58.0 or primer.Tm > 59.5:\n",
    "            isGood = False\n",
    "        if float(primer.Homodimer.dg) < -10000.0:\n",
    "            isGood = False\n",
    "        if float(primer.Hairpin.dg) < -5000.0:\n",
    "            isGood = False\n",
    "    \n",
    "        #return the shortest (18bp ≤) desireable primers\n",
    "        if isGood == True:\n",
    "            #screening returned a good primer; return the primer\n",
    "            if noncanonicalStartCodon == True:\n",
    "                primer = DNASequence(\"atg\".upper() + primer.sequence)\n",
    "            return primer\n",
    "        else:\n",
    "            #primer failed screening, try again with a larger primer using recursion\n",
    "            if whichPrimer == \"F\":\n",
    "                primer = screenPrimer(insert, \"F\", primerSize + 1) \n",
    "            elif whichPrimer == \"R\":\n",
    "                primer = screenPrimer(insert, \"R\", primerSize + 1)\n",
    "            return primer\n",
    "    else:\n",
    "        #parameters are too stringent; no primers were found that matched the criteria given. Return an empty string\n",
    "        return DNASequence(\"\")\n",
    "def addOverhangs(primer1, primer2, orientation, leftBackbone, rightBackbone):\n",
    "    #decide how the construct should be oriented in the final vector and add the overhangs so Gibson will add correctly.\n",
    "    if orientation == \"forward\":\n",
    "        primer2.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer2.sequence\n",
    "        primer1.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer1.sequence\n",
    "        return primer1, primer2\n",
    "    elif orientation == \"reverse\":\n",
    "        primer1.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer1.sequence\n",
    "        primer2.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer2.sequence\n",
    "        return primer1, primer2\n",
    "    else:\n",
    "        print (\"please enter either a forward or reverse strand orientation\")\n",
    "def formatSeq(oldSeq):\n",
    "    if not type(oldSeq) is str:\n",
    "        oldSeq = \"\"\n",
    "    new = oldSeq.replace(\"\\n\", \"\")\n",
    "    for number in range(0,10):\n",
    "        temp = new.replace(str(number), \"\")\n",
    "        new = temp\n",
    "    oldSeq = new\n",
    "    new = oldSeq.replace(\" \", \"\")\n",
    "    return new.upper()\n",
    "\n",
    "def appendPrimers(data, insert, tag, primerList, nameList):\n",
    "    name = str(data.ix[number,\"Protein Name\"]).replace(\" \", \"\") \n",
    "    nameFOR = name + \" FOR \" + tag + \"-tagged\"\n",
    "    nameREV = name + \" REV \" + tag + \"-tagged\"\n",
    "    primerList.append(screenPrimers(insert, leftbackbone , rightbackbone)[0])\n",
    "    nameList.append(nameFOR)\n",
    "    primerList.append(screenPrimers(insert, leftbackbone , rightbackbone)[1])\n",
    "    nameList.append(nameREV)\n",
    "def unpackPrimers(nameList, seqList, finNameList, finSeqList):\n",
    "    for names in nameList:\n",
    "        finNameList.append(names)\n",
    "    for primers in seqList:\n",
    "        finSeqList.append(primers)\n",
    "def createVector(insert, tag, name):\n",
    "    seqPrimerList = []\n",
    "    seqPrimerNameList = []\n",
    "    insert = insert.sequence\n",
    "    annotations = []\n",
    "    pBR322_origin = \"cgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatct\".upper()\n",
    "    kanR2 = \"atgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaa\".upper()\n",
    "    rop = \"gtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctga\".upper()\n",
    "    t7Terminator = \"atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcag\".upper()\n",
    "    tet300 = \"gccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagag\".upper()\n",
    "    pGEX_3_primer = \"cctctgacacatgcagctcccgg\".upper()\n",
    "    thrombin = thrombin = \"gctgccgcgcggcaccag\".upper()\n",
    "    t7promoter = \"cctatagtgagtcgtatta\".upper()\n",
    "    lacO = \"ggggaattgttatccgctcacaattccc\".upper()\n",
    "    hisTag = \"gtgatgatgatgatgatg\".upper()\n",
    "    seqRev = \"gcctatatcgccgacatcacc\".upper()\n",
    "    seqFor = \"tccggatatagttcctcctttcagc\".upper()\n",
    "    rbs = \"ctccttcttaaagttaa\".upper()\n",
    "    hisPlusStop = \"tcagtggtggtggtggtggtg\".upper()\n",
    "    lacI = \"gtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtga\".upper()\n",
    "    tet576 = \"cccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatg\".upper()\n",
    "    \n",
    "    \n",
    "    if tag ==\"C\":\n",
    "        leftBackbone = \"cgcgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgccaatccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg\".upper()\n",
    "        rightBackbone = \"ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcaggg\".upper()\n",
    "        sequence = leftBackbone + reverseComplement(insert[:len(insert) - 3]) + rightBackbone\n",
    "        endLen = len(leftBackbone) + len(insert) - 3\n",
    "        addAnnotation(annotations, reverseComplement(insert[:len(insert) - 3]), sequence, name, 'gene', \"#0000ff\")\n",
    "        \n",
    "    if tag ==\"N\":\n",
    "        leftBackbone = \"cgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgccaatccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatc\".upper()\n",
    "        rightBackbone = \"atggctgccgcgcggcaccaggccgctgctgtgatgatgatgatgatggctgctgcccatggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcg\".upper()\n",
    "        sequence = leftBackbone + reverseComplement(insert) + rightBackbone\n",
    "        endLen = len(leftBackbone) + len(insert)\n",
    "        addAnnotation(annotations, reverseComplement(insert), sequence, name, 'gene', \"#0000ff\")\n",
    "        \n",
    "    offName = str(name) + \" \" + tag + \"-tagged\"\n",
    "    \n",
    "    beta = findSeqPrimers(DNASequence(seqFor), DNASequence(reverseComplement(seqRev)), DNASequence(sequence))\n",
    "    count = 0\n",
    "    for alpha in beta:\n",
    "        count += 1\n",
    "        nom = \"seqFOR\" + str(count) + \" \" + offName\n",
    "        addAnnotation(annotations, alpha, sequence, nom, \"primer\", \"#f2c22e\")\n",
    "        seqPrimerNameList.append(nom)\n",
    "        seqPrimerList.append(alpha)\n",
    "    \n",
    "    addAnnotation(annotations, pBR322_origin, sequence, \"pBR322 Origin\", \"origin\", \"#9faa9a\")\n",
    "    addAnnotation(annotations, kanR2, sequence, \"Kan R2\", \"gene\", \"#359797\")\n",
    "    addAnnotation(annotations, t7Terminator, sequence, \"T7 Terminator\", \"terminator\", \"#d24b19\")\n",
    "    addAnnotation(annotations, tet300, sequence, \"tet (300-563)\", \"gene\", \"#ffe2b9\")\n",
    "    addAnnotation(annotations, pGEX_3_primer, sequence, \"pGEX_3_primer\", \"primer\", \"#5b637e\")\n",
    "    addAnnotation(annotations, thrombin, sequence, \"thrombin\", \"tag\", \"#5f631d\")\n",
    "    addAnnotation(annotations, t7promoter, sequence, \"T7 Promoter\", \"promoter\", \"#c8498b\")\n",
    "    addAnnotation(annotations, lacO, sequence, \"lacO\", \"regulatory\", \"#fffbaa\")\n",
    "    addAnnotation(annotations, lacO, sequence, \"lacI\", \"regulatory\", \"#c3dff3\")\n",
    "    addAnnotation(annotations, hisTag, sequence, \"His Tag\", \"tag\", \"#f9c5fe\")\n",
    "    addAnnotation(annotations, seqFor, sequence, \"pET28+ Forward Primer\", \"primer\", \"#fffafa\")\n",
    "    addAnnotation(annotations, seqRev, sequence, \"pET28+ Reverse Primer\", \"primer\", \"#fffafa\")\n",
    "    addAnnotation(annotations, rbs, sequence, \"T7 Ribosome Binding Site\", \"feature\", \"#e12222\")\n",
    "    addAnnotation(annotations, hisPlusStop, sequence, \"His Tag and Stop Codon\", \"feature\", \"#ddb34e\")\n",
    "    addAnnotation(annotations, tet576, sequence, \"tet(576-851)\", \"feature\", \"#ba9231\")\n",
    "    addAnnotation(annotations, rop, sequence, \"ROP\", \"feature\", \"#1ebdac\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    data = {\n",
    "      \"aliases\": [],\n",
    "      \"annotations\": annotations,\n",
    "      \"bases\": sequence,\n",
    "      \"circular\": True,\n",
    "      \"description\": \"program generated\",\n",
    "      \"folder\": \"lib_iP53B30G\",\n",
    "      \"name\": offName,\n",
    "      \"tags\": []\n",
    "    }\n",
    "\n",
    "    key = 'sk_gnGGnlq0laAcbBcI9yVEIqpE7opJQ'\n",
    "    resp = requests.post('https://api.benchling.com/v1/sequences/', json=data, auth=(key, ''))\n",
    "    print(resp)\n",
    "    return (seqPrimerNameList, seqPrimerList)\n",
    "\n",
    "def findSequence(sequence, plasmid):\n",
    "    if sequence in plasmid:\n",
    "        a = plasmid.find(sequence)\n",
    "        #print(a, a + len(sequence))\n",
    "    else:\n",
    "        return (-1, -1)\n",
    "    return (a, a + len(sequence))\n",
    "def addAnnotation(annotations, sequence, plasmid, name, annotationType, color):\n",
    "    annotation = {\n",
    "          \"color\": color,\n",
    "          \"end\": findSequence(sequence, plasmid)[1],\n",
    "          \"name\": name,\n",
    "          \"start\": findSequence(sequence, plasmid)[0],\n",
    "          \"strand\": 0,\n",
    "          \"type\": annotationType\n",
    "    }\n",
    "    if findSequence(sequence, plasmid) != (-1, -1):\n",
    "        annotations.append(annotation)\n",
    "    return annotations\n",
    "\n",
    "primerList = []\n",
    "nameList = []\n",
    "\n",
    "for number in range(0,49):\n",
    "    tag = \"\"\n",
    "    if (data.ix[number, \"His Tagged Where?\"]) == \"C\":\n",
    "        tag = \"C\"\n",
    "        #Correct the backbone so that the primers generated add in the correct area of the backbone\n",
    "        leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg'.upper())\n",
    "        rightbackbone = DNASequence('ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcgcgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgcca'.upper())\n",
    "        insert = DNASequence(formatSeq(data.ix[number,\"Nucleotide Sequence\"]))\n",
    "        name = data.ix[number,\"Protein Name\"]\n",
    "        a = createVector(insert, tag, name)\n",
    "        unpackPrimers(a[0], a[1], seqPrimerNames, seqPrimers)\n",
    "        insert = DNASequence(insert.sequence[:len(insert.sequence) - 3])\n",
    "        if len(insert.sequence) != 0:\n",
    "            appendPrimers(data, insert, tag, primerList, nameList)\n",
    "    elif (data.ix[number, \"His Tagged Where?\"]) == \"N\": \n",
    "        tag = \"N\"\n",
    "        #Correct the backbone so that the primers generated add in the correct area of the backbone\n",
    "        leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatc'.upper())\n",
    "        rightbackbone = DNASequence('atggctgccgcgcggcaccaggccgctgctgtgatgatgatgatgatggctgctgcccatggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccag'.upper())\n",
    "        insert = DNASequence(formatSeq(data.ix[number,\"Nucleotide Sequence\"]))\n",
    "        name = data.ix[number,\"Protein Name\"]\n",
    "        a = createVector(insert, tag, name)\n",
    "        unpackPrimers(a[0], a[1], seqPrimerNames, seqPrimers)\n",
    "        if len(insert.sequence) != 0:\n",
    "            appendPrimers(data, insert, tag, primerList, nameList)\n",
    "    else: #We don't know whether unspecified proteins are N or C His-tagged, so we assume they could be either.\n",
    "        leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg'.upper())\n",
    "        rightbackbone = DNASequence('ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcgcgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgcca'.upper())\n",
    "        insert = DNASequence(formatSeq(data.ix[number,\"Nucleotide Sequence\"]))\n",
    "        if len(insert.sequence) != 0:\n",
    "            tag = \"C\"\n",
    "            insert = DNASequence(insert.sequence[:len(insert.sequence) - 3])\n",
    "            name = data.ix[number,\"Protein Name\"]\n",
    "            a = createVector(insert, tag, name)\n",
    "            appendPrimers(data, insert, tag, primerList, nameList)\n",
    "            unpackPrimers(a[0], a[1], seqPrimerNames, seqPrimers)\n",
    "        leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatc'.upper())\n",
    "        rightbackbone = DNASequence('atggctgccgcgcggcaccaggccgctgctgtgatgatgatgatgatggctgctgcccatggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccag'.upper())\n",
    "        insert = DNASequence(formatSeq(data.ix[number,\"Nucleotide Sequence\"]))\n",
    "        if len(insert.sequence) != 0:\n",
    "            tag = \"N\"\n",
    "            name = data.ix[number,\"Protein Name\"]\n",
    "            a = createVector(insert, tag, name)\n",
    "            appendPrimers(data, insert, tag, primerList, nameList)\n",
    "            unpackPrimers(a[0], a[1], seqPrimerNames, seqPrimers)\n",
    "        \n",
    "dataframe = pandas.DataFrame(primerList, nameList)\n",
    "df = pandas.DataFrame(seqPrimers, seqPrimerNames)\n",
    "\n",
    "writer = pandas.ExcelWriter('primerList.xlsx', engine='openpyxl')\n",
    "writer2 = pandas.ExcelWriter('seqPrimerList.xlsx', engine='openpyxl')\n",
    "\n",
    "dataframe.to_excel(writer)\n",
    "df.to_excel(writer2)\n",
    "\n",
    "writer.save()\n",
    "writer2.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGAGAAGCTGGCGCTGAAGTTGAAGT\n",
      "TCAGGCTGCAACCGGAAGGG\n"
     ]
    }
   ],
   "source": [
    "a = screenPrimer(DNASequence('ggaagaagctggcgctgaagttgaagttaaataagccaacccttccggttgcagcctga'.upper()), 'F')\n",
    "print(a.sequence)\n",
    "\n",
    "b = screenPrimer(DNASequence('ggaagaagctggcgctgaagttgaagttaaataagccaacccttccggttgcagcctga'.upper()), 'R')\n",
    "print(b.sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCAACTCAGCTTCCTTTCGGGCTTTGTTAGCAGCCGGATCCTATGCAAACGCGAAGTCCGACTCTAAGATG\n"
     ]
    }
   ],
   "source": [
    "leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatc'.upper())\n",
    "rightbackbone = DNASequence('atggctgccgcgcggcaccaggccgctgctgtgatgatgatgatgatggctgctgcccatggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccag'.upper())\n",
    "\n",
    "a = addOverhangs(DNASequence(\"atgaacacgattaacatcgctaagaacga\".upper()), DNASequence(\"CTATGCAAACGCGAAGTCCGACTCTAAGATG\"), 'reverse', leftbackbone, rightbackbone)\n",
    "\n",
    "print(a[1].sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGGCTAAGCAAGATTATTACGAGATTTTAGGCGTTTCCAAAACAGCGGAAGAGCGTGAAATCAGAAAGGCCTACAAACGCCTGGCCATGAAATACCACCCGGACCGTAACCAGGGTGACAAAGAGGCCGAGGCGAAATTTAAAGAGATCAAGGAAGCTTATGAAGTTCTGACCGACTCGCAAAAACGTGCGGCATACGATCAGTATGGTCATGCTGCGTTTGAGCAAGGTGGCATGGGCGGCGGCGGTTTTGGCGGCGGCGCAGACTTCAGCGATATTTTTGGTGACGTTTTCGGCGATATTTTTGGCGGCGGACGTGGTCGTCAACGTGCGGCGCGCGGTGCTGATTTACGCTATAACATGGAGCTCACCCTCGAAGAAGCTGTACGTGGCGTGACCAAAGAGATCCGCATTCCGACTCTGGAAGAGTGTGACGTTTGCCACGGTAGCGGTGCAAAACCAGGTACACAGCCGCAGACTTGTCCGACCTGTCATGGTTCTGGTCAGGTGCAGATGCGCCAGGGATTCTTCGCTGTACAGCAGACCTGTCCACACTGTCAGGGCCGCGGTACGCTGATCAAAGATCCGTGCAACAAATGTCATGGTCATGGTCGTGTTGAGCGCAGCAAAACGCTGTCCGTTAAAATCCCGGCAGGGGTGGACACTGGAGACCGCATCCGTCTTGCGGGCGAAGGTGAAGCGGGCGAGCATGGCGCACCGGCAGGCGATCTGTACGTTCAGGTTCAGGTTAAACAGCACCCGATTTTCGAGCGTGAAGGCAACAACCTGTATTGCGAAGTCCCGATCAACTTCGCTATGGCGGCGCTGGGTGGCGAAATCGAAGTACCGACCCTTGATGGTCGCGTCAAACTGAAAGTGCCTGGCGAAACCCAGACCGGTAAGCTATTCCGTATGCGCGGTAAAGGCGTCAAGTCTGTCCGCGGTGGCGCACAGGGTGATTTGCTGTGCCGCGTTGTCGTCGAAACACCGGTAGGCCTGAACGAAAGGCAGAAACAGCTGCTGCAAGAGCTGCAAGAAAGCTTCGGTGGCCCAACCGGCGAGCACAACAGCCCGCGCTCAAAGAGCTTCTTTGATGGTGTGAAGAAGTTTTTTGACGACCTGACCCGCTAA\n",
      "\n",
      "\n",
      "0\n",
      "ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTGTAGCGATTATGGATGGCACCACTCCTCGCGTGCTGGAGAACGCCGAAGGCGATCGCACCACGCCTTCTATCATTGCCTATACCCAGGATGGTGAAACTCTAGTTGGTCAGCCGGCTAAACGTCAGGCAGTGACGAACCCGCAAAACACTCTGTTTGCGATTAAACGCCTGATTGGTCGCCGCTTCCAGGACGAAGAAGTACAGCGTGATGTTTCCATCATGCCGTTCAAAATTATTGCTGCTGATAACGGCGACGCATGGGTCGAAGTTAAAGGCCAGAAAATGGCACCGCCGCAGATTTCTGCTGAAGTGCTGAAAAAAATGAAGAAAACCGCTGAAGATTACCTGGGTGAACCGGTAACTGAAGCTGTTATCACCGTACCGGCATACTTTAACGATGCTCAGCGTCAGGCAACCAAAGACGCAGGCCGTATCGCTGGTCTGGAAGTAAAACGTATCATCAACGAACCGACCGCAGCTGCGCTGGCTTACGGTCTGGACAAAGGCACTGGCAACCGTACTATCGCGGTTTATGACCTGGGTGGTGGTACTTTCGATATTTCTATTATCGAAATCGACGAAGTTGACGGCGAAAAAACCTTCGAAGTTCTGGCAACCAACGGTGATACCCACCTGGGGGGTGAAGACTTCGACAGCCGTCTGATCAACTATCTGGTTGAAGAATTCAAGAAAGATCAGGGCATTGACCTGCGCAACGATCCGCTGGCAATGCAGCGCCTGAAAGAAGCGGCAGAAAAAGCGAAAATCGAACTGTCTTCCGCTCAGCAGACCGACGTTAACCTGCCATACATCACTGCAGACGCGACCGGTCCGAAACACATGAACATCAAAGTGACTCGTGCGAAACTGGAAAGCCTGGTTGAAGATCTGGTAAACCGTTCCATTGAGCCGCTGAAAGTTGCACTGCAGGACGCTGGCCTGTCCGTATCTGATATCGACGACGTTATCCTCGTTGGTGGTCAGACTCGTATGCCAATGGTTCAGAAGAAAGTTGCTGAGTTCTTTGGTAAAGAGCCGCGTAAAGACGTTAACCCGGACGAAGCTGTAGCAATCGGTGCTGCTGTTCAGGGTGGTGTTCTGACTGGTGACGTAAAAGACGTACTGCTGCTGGACGTTACCCCGCTGTCTCTGGGTATCGAAACCATGGGCGGTGTGATGACGACGCTGATCGCGAAAAACACCACTATCCCGACCAAGCACAGCCAGGTGTTCTCTACCGCTGAAGACAACCAGTCTGCGGTAACCATCCATGTGCTGCAGGGTGAACGTAAACGTGCGGCTGATAACAAATCTCTGGGTCAGTTCAACCTAGATGGTATCAACCCGGCACCGCGCGGCATGCCGCAGATCGAAGTTACCTTCGATATCGATGCTGACGGTATCCTGCACGTTTCCGCGAAAGATAAAAACAGCGGTAAAGAGCAGAAGATCACCATCAAGGCTTCTTCTGGTCTGAACGAAGATGAAATCCAGAAAATGGTACGCGACGCAGAAGCTAACGCCGAAGCTGACCGTAAGTTTGAAGAGCTGGTACAGACTCGCAACCAGGGCGACCATCTGCTGCACAGCACCCGTAAGCAGGTTGAAGAAGCAGGCGACAAACTGCCGGCTGACGACAAAACTGCTATCGAGTCTGCGCTGACTGCACTGGAAACTGCTCTGAAAGGTGAAGACAAAGCCGCTATCGAAGCGAAAATGCAGGAACTGGCACAGGTTTCCCAGAAACTGATGGAAATCGCCCAGCAGCAACATGCCCAGCAGCAGACTGCCGGTGCTGATGCTTCTGCAAACAACGCGAAAGATGACGATGTTGTCGACGCTGAATTTGAAGAAGTCAAAGACAAAAAATAA\n",
      "\n",
      "\n",
      "1\n",
      "ATGAGTAGTAAAGAACAGAAAACGCCTGAGGGGCAAGCCCCGGAAGAAATTATCATGGATCAGCACGAAGAGATTGAGGCAGTTGAGCCAGAAGCTTCTGCTGAGCAGGTGGATCCGCGCGATGAAAAAGTTGCGAATCTCGAAGCTCAGCTGGCTGAAGCCCAGACCCGTGAACGTGACGGCATTTTGCGTGTAAAAGCCGAAATGGAAAACCTGCGTCGTCGTACTGAACTGGATATTGAAAAAGCCCACAAATTCGCGCTGGAGAAATTCATCAACGAATTGCTGCCGGTGATTGATAGCCTGGATCGTGCGCTGGAAGTGGCTGATAAAGCTAACCCGGATATGTCTGCGATGGTTGAAGGCATTGAGCTGACGCTGAAGTCGATGCTGGATGTTGTGCGTAAGTTTGGCGTTGAAGTGATCGCCGAAACTAACGTCCCACTGGACCCGAATGTGCATCAGGCCATCGCAATGGTGGAATCTGATGACGTTGCGCCAGGTAACGTACTGGGCATTATGCAGAAGGGTTATACGCTGAATGGTCGTACGATTCGTGCGGCGATGGTTACTGTAGCGAAAGCAAAAGCTTAA\n",
      "\n",
      "\n",
      "2\n",
      "ATGCGTCTGGATCGTCTTACTAATAAATTCCAGCTTGCTCTTGCCGATGCCCAATCACTTGCACTCGGGCACGACAACCAATTTATCGAACCACTTCATTTAATGAGCGCCCTGCTGAATCAGGAAGGGGGTTCGGTTAGTCCTTTATTAACATCCGCTGGCATAAATGCTGGCCAGTTGCGCACAGATATCAATCAGGCATTAAATCGTTTACCGCAGGTTGAAGGTACTGGTGGTGATGTCCAGCCATCACAGGATCTGGTGCGCGTTCTTAATCTTTGCGACAAGCTGGCGCAAAAACGTGGTGATAACTTTATCTCGTCAGAACTGTTCGTTCTGGCGGCACTTGAGTCTCGCGGCACGCTGGCCGACATCCTGAAAGCAGCAGGGGCGACCACCGCCAACATTACTCAAGCGATTGAACAAATGCGTGGAGGTGAAAGCGTGAACGATCAAGGTGCTGAAGACCAACGTCAGGCTTTGAAAAAATATACCATCGACCTTACCGAACGAGCCGAACAGGGCAAACTCGATCCGGTGATTGGTCGTGATGAAGAAATTCGCCGTACCATTCAGGTGCTGCAACGTCGTACTAAAAATAACCCGGTACTGATTGGTGAACCCGGCGTCGGTAAAACTGCCATCGTTGAAGGTCTGGCGCAGCGTATTATCAACGGCGAAGTGCCGGAAGGGTTGAAAGGCCGCCGGGTACTGGCGCTGGATATGGGCGCGCTGGTGGCTGGGGCGAAATATCGCGGTGAGTTTGAAGAACGTTTAAAAGGCGTGCTTAACGATCTTGCCAAACAGGAAGGCAACGTCATCCTATTTATCGACGAATTACATACCATGGTCGGCGCGGGTAAAGCCGATGGCGCAATGGACGCCGGAAACATGCTGAAACCGGCGCTGGCGCGTGGTGAATTGCACTGCGTAGGTGCCACGACGCTTGACGAATATCGCCAGTACATTGAAAAAGATGCTGCGCTGGAACGTCGTTTCCAGAAAGTGTTTGTTGCCGAGCCTTCTGTTGAAGATACCATTGCGATTCTGCGTGGCCTGAAAGAACGTTACGAATTGCACCACCATGTGCAAATTACTGACCCGGCAATTGTTGCAGCGGCGACGTTGTCTCATCGCTACATTGCTGACCGTCAGCTGCCGGATAAAGCCATCGACCTGATCGATGAAGCAGCATCCAGCATTCGTATGCAGATTGACTCAAAACCAGAAGAACTCGACCGACTCGATCGTCGTATCATCCAGCTCAAACTGGAACAACAGGCGTTAATGAAAGAGTCTGATGAAGCCAGTAAAAAACGTCTGGATATGCTCAACGAAGAACTGAGCGACAAAGAACGTCAGTACTCCGAGTTAGAAGAAGAGTGGAAAGCAGAGAAGGCATCGCTTTCTGGTACGCAGACCATTAAAGCGGAACTGGAACAGGCGAAAATCGCTATTGAACAGGCTCGCCGTGTGGGGGACCTGGCGCGGATGTCTGAACTGCAATACGGCAAAATCCCGGAACTGGAAAAGCAACTGGAAGCCGCAACGCAGCTCGAAGGCAAAACTATGCGTCTGTTGCGTAATAAAGTGACCGACGCCGAAATTGCTGAAGTGCTGGCGCGTTGGACGGGGATTCCGGTTTCTCGCATGATGGAAAGCGAGCGCGAAAAACTGCTGCGTATGGAGCAAGAACTGCACCATCGCGTAATTGGTCAGAACGAAGCGGTTGATGCGGTATCTAACGCTATTCGTCGTAGCCGTGCGGGGCTGGCGGATCCAAATCGCCCGATTGGTTCATTCCTGTTCCTCGGCCCAACTGGTGTGGGGAAAACAGAGCTTTGTAAGGCGCTGGCGAACTTTATGTTTGATAGCGACGAGGCGATGGTCCGTATCGATATGTCCGAGTTTATGGAGAAACACTCGGTGTCTCGTTTGGTTGGTGCGCCTCCGGGATATGTCGGTTATGAAGAAGGTGGCTACCTGACCGAAGCGGTGCGTCGTCGTCCGTATTCCGTCATCCTGCTGGATGAAGTGGAAAAAGCGCATCCGGATGTCTTCAACATTCTGTTGCAGGTACTGGATGATGGGCGTCTGACTGACGGGCAAGGGAGAACGGTCGACTTCCGTAATACGGTCGTCATTATGACCTCTAACCTCGGTTCCGATCTGATTCAGGAACGCTTCGGTGAACTGGATTATGCGCACATGAAAGAGCTGGTGCTCGGTGTGGTAAGCCATAACTTCCGTCCGGAATTCATTAACCGTATCGATGAAGTGGTGGTCTTCCATCCGCTGGGTGAACAGCACATTGCCTCGATTGCGCAGATTCAGTTGAAACGTCTGTACAAACGTCTGGAAGAACGTGGTTATGAAATCCACATTTCTGACGAGGCGCTGAAACTGCTGAGCGAGAACGGTTACGATCCGGTCTATGGTGCACGTCCTCTGAAACGTGCAATTCAGCAGCAGATCGAAAACCCGCTGGCACAGCAAATACTGTCTGGTGAATTGGTTCCGGGTAAAGTGATTCGCCTGGAAGTTAATGAAGACCGGATTGTCGCCGTCCAGTAA\n",
      "\n",
      "\n",
      "3\n",
      "ATGAAAGGACAAGAAACTCGTGGTTTTCAGTCAGAAGTGAAACAGCTTCTGCACCTGATGATCCATTCTCTCTATTCCAATAAAGAAATCTTCCTGCGTGAGCTTATCTCTAACGCCTCCGATGCGGCGGACAAGCTGCGTTTCCGTGCGCTCTCTAACCCGGACCTGTACGAAGGTGATGGCGAACTACGCGTTCGTGTCTCTTTCGATAAAGACAAGCGTACGCTGACCATCTCCGATAACGGCGTGGGGATGACCCGCGACGAAGTGATTGACCATCTGGGGACTATCGCTAAATCCGGTACCAAATCATTCCTCGAATCCCTGGGTTCTGACCAGGCGAAAGACAGCCAGCTGATCGGTCAGTTTGGTGTTGGTTTCTACTCTGCGTTTATCGTGGCCGACAAAGTGACCGTGCGTACTCGCGCGGCAGGCGAAAAACCAGAAAATGGCGTCTTCTGGGAATCGGCTGGCGAAGGTGAATACACCGTTGCCGACATCACCAAAGAAGATCGTGGTACTGAAATCACCCTGCATCTGCGTGAAGGCGAAGACGAGTTCCTCGATGACTGGCGCGTGCGTTCCATCATCAGCAAATACTCCGACCATATCGCGCTGCCGGTAGAGATCGAAAAACGCGAAGAGAAAGACGGCGAAACCGTTATCTCCTGGGAGAAAATCAACAAAGCGCAGGCGCTGTGGACTCGTAACAAGTCGGAAATCACCGATGAAGAGTACAAAGAGTTCTACAAACACATCGCCCACGACTTTAATGATCCGCTGACCTGGAGCCACAACCGTGTTGAAGGTAAGCAGGAGTACACCAGCCTGCTGTACATCCCGTCCCAGGCTCCGTGGGATATGTGGAACCGCGATCATAAACACGGCCTGAAACTGTATGTTCAGCGTGTGTTCATCATGGACGACGCAGAACAGTTCATGCCGAACTATCTGCGCTTCGTGCGTGGTCTGATTGACTCCAGCGATCTGCCGCTGAACGTTTCCCGTGAAATCCTCCAGGACAGCACGGTAACGCGTAACCTGCGCAATGCGCTGACCAAGCGTGTGCTGCAAATGCTGGAAAAACTGGCGAAAGACGACGCGGAAAAATACCAGACCTTCTGGCAACAGTTTGGCCTGGTACTGAAAGAAGGTCCGGCGGAAGATTTCGCTAACCAGGAAGCGATCGCCAAACTGCTGCGTTTTGCTTCTACCCATACCGATTCTTCTGCGCAGACCGTATCTCTGGAAGACTACGTTTCCCGCATGAAAGAAGGGCAGGAGAAAATCTACTACATCACCGCAGACAGCTATGCGGCAGCGAAGAGCAGCCCGCACCTGGAACTGCTGCGTAAGAAAGGCATCGAAGTTCTGCTGCTTTCCGACCGCATCGATGAGTGGATGATGAACTATCTGACTGAGTTCGACGGTAAACCGTTCCAGTCGGTGTCTAAAGTTGACGAGTCGCTTGAAAAACTGGCTGACGAAGTTGATGAGAGCGCGAAAGAAGCGGAGAAAGCACTGACTCCGTTCATCGACCGTGTGAAAGCCCTGCTCGGCGAGCGCGTGAAAGATGTCCGTCTGACTCACCGTCTGACCGATACGCCAGCGATCGTTTCGACCGACGCGGACGAAATGAGCACTCAGATGGCGAAACTGTTCGCTGCGGCGGGCCAGAAAGTGCCAGAAGTGAAATACATCTTCGAACTGAACCCGGATCACGTACTGGTGAAACGTGCGGCAGATACTGAAGATGAAGCGAAGTTCAGCGAGTGGGTAGAACTGCTGCTGGATCAGGCGCTGCTGGCAGAACGCGGCACGCTGGAAGATCCGAACCTGTTTATTCGTCGTATGAACCAGCTGCTGGTTTCCTGA\n",
      "\n",
      "\n",
      "4\n",
      "ATGTCAGAACAAAACAACACTGAAATGACTTTCCAGATCCAACGTATTTATACCAAGGATATCTCTTTCGAAGCGCCGAACGCGCCGCACGTTTTCCAGAAAGATTGGCAACCAGAAGTTAAACTTGATCTGGATACGGCATCTTCCCAACTGGCAGATGACGTATACGAAGTGGTACTGCGTGTTACCGTAACGGCCTCTTTGGGCGAAGAAACCGCGTTCCTGTGTGAAGTTCAGCAGGGCGGTATTTTCTCCATCGCGGGTATCGAAGGCACCCAGATGGCGCATTGCCTGGGAGCATACTGCCCGAACATTCTGTTCCCGTATGCTCGTGAGTGCATCACCAGCATGGTATCCCGCGGTACATTCCCGCAACTGAACCTTGCGCCGGTTAACTTCGATGCGCTGTTCATGAACTATTTGCAGCAGCAGGCTGGCGAAGGTACTGAAGAACATCAGGATGCCTGA\n",
      "\n",
      "\n",
      "5\n",
      "ATGCAAGTTTCAGTTGAAACCACTCAAGGCCTTGGCCGCCGTGTAACGATTACTATCGCTGCTGACAGCATCGAGACCGCTGTTAAAAGCGAGCTGGTCAACGTTGCGAAAAAAGTACGTATTGACGGCTTCCGCAAAGGCAAAGTGCCAATGAATATCGTTGCTCAGCGTTATGGCGCGTCTGTACGCCAGGACGTTCTGGGTGACCTGATGAGCCGTAACTTCATTGACGCCATCATTAAAGAAAAAATCAATCCGGCTGGCGCACCGACTTATGTTCCGGGCGAATACAAGCTGGGTGAAGACTTCACTTACTCTGTAGAGTTTGAAGTTTATCCGGAAGTTGAACTGCAGGGTCTGGAAGCGATCGAAGTTGAAAAACCGATCGTTGAAGTGACCGACGCTGACGTTGACGGCATGCTGGATACTCTGCGTAAACAGCAGGCGACCTGGAAAGAAAAAGACGGCGCTGTTGAAGCAGAAGACCGCGTAACCATCGACTTCACCGGTTCTGTAGACGGCGAAGAGTTCGAAGGCGGTAAAGCGTCTGATTTCGTACTGGCGATGGGCCAGGGTCGTATGATCCCGGGCTTTGAAGACGGTATCAAAGGCCACAAAGCTGGCGAAGAGTTCACCATCGACGTGACCTTCCCGGAAGAATACCACGCAGAAAACCTGAAAGGTAAAGCAGCGAAATTCGCTATCAACCTGAAGAAAGTTGAAGAGCGTGAACTGCCGGAACTGACTGCAGAATTCATCAAACGTTTCGGCGTTGAAGATGGTTCCGTAGAAGGTCTGCGCGCTGAAGTGCGTAAAAACATGGAGCGCGAGCTGAAGAGCGCCATCCGTAACCGCGTTAAGTCTCAGGCGATCGAAGGTCTGGTAAAAGCTAACGACATCGACGTACCGGCTGCGCTGATCGACAGCGAAATCGACGTTCTGCGTCGCCAGGCTGCACAGCGTTTCGGTGGCAACGAAAAACAAGCTCTGGAACTGCCGCGCGAACTGTTCGAAGAACAGGCTAAACGCCGCGTAGTTGTTGGCCTGCTGCTGGGCGAAGTTATCCGCACCAACGAGCTGAAAGCTGACGAAGAGCGCGTGAAAGGCCTGATCGAAGAGATGGCTTCTGCGTACGAAGATCCGAAAGAAGTTATCGAGTTCTACAGCAAAAACAAAGAACTGATGGACAACATGCGCAATGTTGCTCTGGAAGAACAGGCTGTTGAAGCTGTACTGGCGAAAGCGAAAGTGACTGAAAAAGAAACCACTTTCAACGAGCTGATGAACCAGCAGGCGTAA\n",
      "\n",
      "\n",
      "6\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n"
     ]
    }
   ],
   "source": [
    "import primer3\n",
    "import pandas\n",
    "import math\n",
    "sequences = []\n",
    "breakdown = []\n",
    "\n",
    "seqPrimers = []\n",
    "seqPrimerNames = []\n",
    "\n",
    "data = pandas.read_excel(\"/Users/aleclourenco/Desktop/REU-2017-Endy-Lab/PURE Datasheet.xlsx\", sheetname='Alec\\'s Chaperones', skiprows=[0])\n",
    "\n",
    "for number in range (0, 7):\n",
    "    print(formatSeq(data.ix[number,\"DNA Sequence\"]))\n",
    "    print (\"\\n\")\n",
    "    print(number)\n",
    "class DNASequence():\n",
    "    def __init__(self, sequence):\n",
    "        #Initialize the DNA variable, which has sequence info as well as additional properties (Tm, homodimer & hairpin formation)\n",
    "        self.sequence = sequence\n",
    "        self.composition = self.getComposition(sequence)\n",
    "        if len(sequence) != 0:\n",
    "            self.gcContent = self.getGCcontent(sequence)\n",
    "        else:\n",
    "            self.gcContent = 0\n",
    "        if len(sequence) < 60:\n",
    "            self.Tm = primer3.calcTm(sequence)\n",
    "            self.Homodimer = primer3.calcHomodimer(sequence)\n",
    "            self.Hairpin = primer3.calcHairpin(sequence)\n",
    "        else:\n",
    "            self.Tm = 100000 #set the Tm arbitrarily high\n",
    "        \n",
    "    def getComposition(self, sequence):\n",
    "        composition = {\"A\" : 0, \"G\" : 0, \"C\" : 0, \"T\" : 0}\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                composition[\"A\"] = composition[\"A\"] + 1\n",
    "            elif base == \"T\":\n",
    "                composition[\"T\"] = composition[\"T\"] + 1\n",
    "            elif base == \"G\":\n",
    "                composition[\"G\"] = composition[\"G\"] + 1\n",
    "            elif base == \"C\":\n",
    "                composition[\"C\"] = composition[\"C\"] + 1\n",
    "            else:\n",
    "                print(\"error 2\")\n",
    "        return composition\n",
    "    def getGCcontent(self, sequence):\n",
    "        GCpercent = (self.composition[\"G\"] + self.composition[\"C\"])/(self.composition[\"G\"] + self.composition[\"T\"] + self.composition[\"A\"] + self.composition[\"C\"])\n",
    "        return GCpercent\n",
    "    def reverseComplement(self, sequence):\n",
    "        #Takes the complement of the sequence and reverses it.\n",
    "        reverseComplement = \"\"\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                reverseComplement = \"T\" + reverseComplement\n",
    "            elif base == \"T\":\n",
    "                reverseComplement = \"A\" + reverseComplement\n",
    "            elif base == \"G\":\n",
    "                reverseComplement = \"C\" + reverseComplement\n",
    "            elif base == \"C\":\n",
    "                reverseComplement = \"G\" + reverseComplement \n",
    "            else:\n",
    "                print (\"error\")\n",
    "        return (reverseComplement)\n",
    "def reverseComplement(sequence):\n",
    "    #Takes the complement of the sequence and reverses it.\n",
    "    reverseComplement = \"\"\n",
    "    for base in sequence:\n",
    "        if base == \"A\":\n",
    "            reverseComplement = \"T\" + reverseComplement \n",
    "        elif base == \"T\":\n",
    "            reverseComplement = \"A\" + reverseComplement\n",
    "        elif base == \"G\":\n",
    "            reverseComplement = \"C\" + reverseComplement\n",
    "        elif base == \"C\":\n",
    "            reverseComplement = \"G\" + reverseComplement\n",
    "        else:\n",
    "            print (\"error\")\n",
    "    return (reverseComplement)\n",
    "def screenPrimers(insert, upstreamBackbone, downstreamBackbone):\n",
    "    #run the main program to create and screen the forward and reverse primers and then add overhangs.\n",
    "    forwardPrimer = screenPrimer(insert, \"F\")\n",
    "    #print(forwardPrimer.sequence, \"forward\")\n",
    "    reversePrimer = screenPrimer(insert, \"R\")\n",
    "    #print(reversePrimer.sequence, \"reverse\")\n",
    "    if len(forwardPrimer.sequence) and len(reversePrimer.sequence) != 0:\n",
    "        primers = addOverhangs(forwardPrimer, reversePrimer, \"reverse\", upstreamBackbone, downstreamBackbone)\n",
    "        return (primers[0].sequence, primers[1].sequence)\n",
    "    else:\n",
    "        return (\"You cannot make primers with the given parameters\", \"You cannot make primers with the given parameters\")\n",
    "    \n",
    "def screenPrimer(insert, whichPrimer, primerSize=18): #Make primers for an insert to amplify it such that it can be assembled with Gibson\n",
    "    if primerSize < 40:\n",
    "        #Make sure you are selecting the right region to prime. \n",
    "        #This algorithm creates primers that amplify the WHOLE SEQUENCE minus the stop codon at the end.\n",
    "        noncanonicalStartCodon = False\n",
    "        if whichPrimer == \"F\": \n",
    "            primer = DNASequence(insert.sequence[:primerSize])\n",
    "            if (primer.sequence[:3].lower() != \"atg\"):\n",
    "                primer = DNASequence(insert.sequence[3:primerSize])\n",
    "                noncanonicalStartCodon = True\n",
    "        elif whichPrimer == \"R\":\n",
    "            primer = DNASequence(reverseComplement(insert.sequence[len(insert.sequence) - primerSize: len(insert.sequence)]))\n",
    "        else:\n",
    "            print(\"error\")\n",
    "        #Add various screens below here. You can update the parameters to be more or less stringent depending on how likely you would like your primers to work\n",
    "        isGood = True\n",
    "        if primer.Tm < 58.0 or primer.Tm > 59.5:\n",
    "            isGood = False\n",
    "        if float(primer.Homodimer.dg) < -10000.0:\n",
    "            isGood = False\n",
    "        if float(primer.Hairpin.dg) < -5000.0:\n",
    "            isGood = False\n",
    "    \n",
    "        #return the shortest (18bp ≤) desireable primers\n",
    "        if isGood == True:\n",
    "            #screening returned a good primer; return the primer\n",
    "            if noncanonicalStartCodon == True:\n",
    "                primer = DNASequence(\"atg\".upper() + primer.sequence)\n",
    "            return primer\n",
    "        else:\n",
    "            #primer failed screening, try again with a larger primer using recursion\n",
    "            if whichPrimer == \"F\":\n",
    "                primer = screenPrimer(insert, \"F\", primerSize + 1) \n",
    "            elif whichPrimer == \"R\":\n",
    "                primer = screenPrimer(insert, \"R\", primerSize + 1)\n",
    "            return primer\n",
    "    else:\n",
    "        #parameters are too stringent; no primers were found that matched the criteria given. Return an empty string\n",
    "        return DNASequence(\"\")\n",
    "def addOverhangs(primer1, primer2, orientation, leftBackbone, rightBackbone):\n",
    "    #decide how the construct should be oriented in the final vector and add the overhangs so Gibson will add correctly.\n",
    "    if orientation == \"forward\":\n",
    "        primer2.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer2.sequence\n",
    "        primer1.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer1.sequence\n",
    "        return primer1, primer2\n",
    "    elif orientation == \"reverse\":\n",
    "        primer1.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer1.sequence\n",
    "        primer2.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer2.sequence\n",
    "        return primer1, primer2\n",
    "    else:\n",
    "        print (\"please enter either a forward or reverse strand orientation\")\n",
    "def formatSeq(oldSeq):\n",
    "    if not type(oldSeq) is str:\n",
    "        oldSeq = \"\"\n",
    "    new = oldSeq.replace(\"\\n\", \"\")\n",
    "    for number in range(0,10):\n",
    "        temp = new.replace(str(number), \"\")\n",
    "        new = temp\n",
    "    oldSeq = new\n",
    "    new = oldSeq.replace(\" \", \"\")\n",
    "    return new.upper()\n",
    "\n",
    "def appendPrimers(data, insert, tag, primerList, nameList):\n",
    "    name = str(data.ix[number,\"Chaperone Name\"]).replace(\" \", \"\") \n",
    "    nameFOR = name + \" FOR \" + tag + \"-tagged\"\n",
    "    nameREV = name + \" REV \" + tag + \"-tagged\"\n",
    "    primerList.append(screenPrimers(insert, leftbackbone , rightbackbone)[0])\n",
    "    nameList.append(nameFOR)\n",
    "    primerList.append(screenPrimers(insert, leftbackbone , rightbackbone)[1])\n",
    "    nameList.append(nameREV)\n",
    "def unpackPrimers(nameList, seqList, finNameList, finSeqList):\n",
    "    for names in nameList:\n",
    "        finNameList.append(names)\n",
    "    for primers in seqList:\n",
    "        finSeqList.append(primers)\n",
    "\n",
    "\n",
    "primerList = []\n",
    "nameList = []\n",
    "\n",
    "for number in range(0,7):\n",
    "    tag = \"\"\n",
    "    #C tag them all for now\n",
    "    leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg'.upper())\n",
    "    rightbackbone = DNASequence('ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcgcgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgcca'.upper())\n",
    "    \n",
    "    insert = DNASequence(formatSeq(data.ix[number,\"DNA Sequence\"]))\n",
    "    insert = DNASequence(insert.sequence[:len(insert.sequence) - 3])\n",
    "    if len(insert.sequence) != 0:\n",
    "        tag = \"C\"\n",
    "        name = str(data.ix[number,\"Chaperone Name\"]).replace(\" \", \"\")\n",
    "        a = createVector(insert, tag, name)\n",
    "        appendPrimers(data, insert, tag, primerList, nameList)\n",
    "        if a != None:\n",
    "            unpackPrimers(a[0], a[1], seqPrimerNames, seqPrimers)\n",
    "        \n",
    "dataframe = pandas.DataFrame(primerList, nameList)\n",
    "df = pandas.DataFrame(seqPrimers, seqPrimerNames)\n",
    "\n",
    "writer = pandas.ExcelWriter('primerList3.xlsx', engine='openpyxl')\n",
    "writer2 = pandas.ExcelWriter('seqPrimerListChap.xlsx', engine='openpyxl')\n",
    "\n",
    "dataframe.to_excel(writer)\n",
    "df.to_excel(writer2)\n",
    "\n",
    "writer.save()\n",
    "writer2.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import datetime as dt\n",
    "import pandas as pd\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "def add_frame_to_workbook(filename, tabname, dataframe, timestamp):\n",
    "    \"\"\"\n",
    "    Save a dataframe to a workbook tab with the filename and tabname\n",
    "    coded to timestamp\n",
    "\n",
    "    :param filename: filename to create, can use strptime formatting\n",
    "    :param tabname: tabname to create, can use strptime formatting\n",
    "    :param dataframe: dataframe to save to workbook\n",
    "    :param timestamp: timestamp associated with dataframe\n",
    "    :return: None\n",
    "    \"\"\"\n",
    "    filename = timestamp.strftime(filename)\n",
    "    sheet_name = timestamp.strftime(tabname)\n",
    "\n",
    "    # create a writer for this month and year\n",
    "    writer = pd.ExcelWriter(filename, engine='openpyxl')\n",
    "\n",
    "    try:\n",
    "        # try to open an existing workbook\n",
    "        writer.book = load_workbook(filename)\n",
    "\n",
    "        # copy existing sheets\n",
    "        writer.sheets = dict(\n",
    "            (ws.title, ws) for ws in writer.book.worksheets)\n",
    "    except IOError:\n",
    "        # file does not exist yet, we will create it\n",
    "        pass\n",
    "\n",
    "    # write out the new sheet\n",
    "    dataframe.to_excel(writer, sheet_name=sheet_name)\n",
    "\n",
    "    # save the workbook\n",
    "    writer.save()\n",
    "\n",
    "data = [x.strip().split() for x in \"\"\"\n",
    "            Primer Name  Sequence\n",
    "                    FOR  ATGTGCCATGA...\n",
    "                    REV  TGCCTAGTCA...\n",
    "\"\"\".split('\\n')[1:-1]]\n",
    "df = pd.DataFrame(data=data[1:], columns=data[0])\n",
    "\n",
    "name_template = './sample-%m-%y.xlsx'\n",
    "tab_template = '%d_%H_%M'\n",
    "now = dt.datetime.now()\n",
    "in_an_hour = now + dt.timedelta(hours=1)\n",
    "add_frame_to_workbook(name_template, tab_template, df, now)\n",
    "add_frame_to_workbook(name_template, tab_template, df, in_an_hour)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "df = pd.DataFrame({'name' : pd.Series(np.random.randn(3)),'Sequence' : pd.Series(np.random.randn(4))})\n",
    "print(df)\n",
    "\n",
    "writer = pd.ExcelWriter('test2.xlsx', engine='openpyxl')\n",
    "\n",
    "df.to_excel(writer)\n",
    "\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import primer3\n",
    "import pandas\n",
    "import math\n",
    "sequences = []\n",
    "breakdown = []\n",
    "data = pandas.read_excel(\"/Users/aleclourenco/Desktop/REU-2017-Endy-Lab/PURE Datasheet.xlsx\", sheetname=\"List of Ribosome Protein Compon\")\n",
    "\n",
    "class DNASequence():\n",
    "    def __init__(self, sequence):\n",
    "        #Initialize the DNA variable, which has sequence info as well as additional properties (Tm, homodimer & hairpin formation)\n",
    "        self.sequence = sequence\n",
    "        self.composition = self.getComposition(sequence)\n",
    "        if len(sequence) != 0:\n",
    "            self.gcContent = self.getGCcontent(sequence)\n",
    "        else:\n",
    "            self.gcContent = 0\n",
    "        if len(sequence) < 60:\n",
    "            self.Tm = primer3.calcTm(sequence)\n",
    "            self.Homodimer = primer3.calcHomodimer(sequence)\n",
    "            self.Hairpin = primer3.calcHairpin(sequence)\n",
    "        else:\n",
    "            self.Tm = 100000 #set the Tm arbitrarily high\n",
    "        \n",
    "    def getComposition(self, sequence):\n",
    "        composition = {\"A\" : 0, \"G\" : 0, \"C\" : 0, \"T\" : 0}\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                composition[\"A\"] = composition[\"A\"] + 1\n",
    "            elif base == \"T\":\n",
    "                composition[\"T\"] = composition[\"T\"] + 1\n",
    "            elif base == \"G\":\n",
    "                composition[\"G\"] = composition[\"G\"] + 1\n",
    "            elif base == \"C\":\n",
    "                composition[\"C\"] = composition[\"C\"] + 1\n",
    "            else:\n",
    "                print(\"error 2\")\n",
    "        return composition\n",
    "    def getGCcontent(self, sequence):\n",
    "        GCpercent = (self.composition[\"G\"] + self.composition[\"C\"])/(self.composition[\"G\"] + self.composition[\"T\"] + self.composition[\"A\"] + self.composition[\"C\"])\n",
    "        return GCpercent\n",
    "    def reverseComplement(self, sequence):\n",
    "        #Takes the complement of the sequence and reverses it.\n",
    "        reverseComplement = \"\"\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                reverseComplement = \"T\" + reverseComplement\n",
    "            elif base == \"T\":\n",
    "                reverseComplement = \"A\" + reverseComplement\n",
    "            elif base == \"G\":\n",
    "                reverseComplement = \"C\" + reverseComplement\n",
    "            elif base == \"C\":\n",
    "                reverseComplement = \"G\" + reverseComplement \n",
    "            else:\n",
    "                print (\"error\")\n",
    "        return (reverseComplement)\n",
    "def reverseComplement(sequence):\n",
    "    #Takes the complement of the sequence and reverses it.\n",
    "    reverseComplement = \"\"\n",
    "    for base in sequence:\n",
    "        if base == \"A\":\n",
    "            reverseComplement = \"T\" + reverseComplement \n",
    "        elif base == \"T\":\n",
    "            reverseComplement = \"A\" + reverseComplement\n",
    "        elif base == \"G\":\n",
    "            reverseComplement = \"C\" + reverseComplement\n",
    "        elif base == \"C\":\n",
    "            reverseComplement = \"G\" + reverseComplement\n",
    "        else:\n",
    "            print (\"error\")\n",
    "    return (reverseComplement)\n",
    "def screenPrimers(insert, upstreamBackbone, downstreamBackbone):\n",
    "    #run the main program to create and screen the forward and reverse primers and then add overhangs.\n",
    "    forwardPrimer = screenPrimer(insert, \"F\")\n",
    "    #print(forwardPrimer.sequence, \"forward\")\n",
    "    reversePrimer = screenPrimer(insert, \"R\")\n",
    "    #print(reversePrimer.sequence, \"reverse\")\n",
    "    if len(forwardPrimer.sequence) and len(reversePrimer.sequence) != 0:\n",
    "        primers = addOverhangs(forwardPrimer, reversePrimer, \"reverse\", upstreamBackbone, downstreamBackbone)\n",
    "        return (primers[0].sequence, primers[1].sequence)\n",
    "    else:\n",
    "        return (\"You cannot make primers with the given parameters\", \"You cannot make primers with the given parameters\")\n",
    "    \n",
    "def screenPrimer(insert, whichPrimer, primerSize=18): #Make primers for an insert to amplify it such that it can be assembled with Gibson\n",
    "    if primerSize < 40:\n",
    "        #Make sure you are selecting the right region to prime. \n",
    "        #This algorithm creates primers that amplify the WHOLE SEQUENCE minus the stop codon at the end.\n",
    "        noncanonicalStartCodon = False\n",
    "        if whichPrimer == \"F\": \n",
    "            primer = DNASequence(insert.sequence[:primerSize])\n",
    "            if (primer.sequence[:3].lower() != \"atg\"):\n",
    "                primer = DNASequence(insert.sequence[3:primerSize])\n",
    "                noncanonicalStartCodon = True\n",
    "        elif whichPrimer == \"R\":\n",
    "            primer = DNASequence(reverseComplement(insert.sequence[len(insert.sequence) - primerSize: len(insert.sequence)]))\n",
    "        else:\n",
    "            print(\"error\")\n",
    "        #Add various screens below here. You can update the parameters to be more or less stringent depending on how likely you would like your primers to work\n",
    "        isGood = True\n",
    "        if primer.Tm < 58.0 or primer.Tm > 59.5:\n",
    "            isGood = False\n",
    "        if float(primer.Homodimer.dg) < -10000.0:\n",
    "            isGood = False\n",
    "        if float(primer.Hairpin.dg) < -5000.0:\n",
    "            isGood = False\n",
    "    \n",
    "        #return the shortest (18bp ≤) desireable primers\n",
    "        if isGood == True:\n",
    "            #screening returned a good primer; return the primer\n",
    "            if noncanonicalStartCodon == True:\n",
    "                primer = DNASequence(\"atg\".upper() + primer.sequence)\n",
    "            return primer\n",
    "        else:\n",
    "            #primer failed screening, try again with a larger primer using recursion\n",
    "            if whichPrimer == \"F\":\n",
    "                primer = screenPrimer(insert, \"F\", primerSize + 1) \n",
    "            elif whichPrimer == \"R\":\n",
    "                primer = screenPrimer(insert, \"R\", primerSize + 1)\n",
    "            return primer\n",
    "    else:\n",
    "        #parameters are too stringent; no primers were found that matched the criteria given. Return an empty string\n",
    "        return DNASequence(\"\")\n",
    "def addOverhangs(primer1, primer2, orientation, leftBackbone, rightBackbone):\n",
    "    #decide how the construct should be oriented in the final vector and add the overhangs so Gibson will add correctly.\n",
    "    if orientation == \"forward\":\n",
    "        primer2.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer2.sequence\n",
    "        primer1.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer1.sequence\n",
    "        return primer1, primer2\n",
    "    elif orientation == \"reverse\":\n",
    "        primer1.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer1.sequence\n",
    "        primer2.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer2.sequence\n",
    "        return primer1, primer2\n",
    "    else:\n",
    "        print (\"please enter either a forward or reverse strand orientation\")\n",
    "def formatSeq(oldSeq):\n",
    "    if not type(oldSeq) is str:\n",
    "        oldSeq = \"\"\n",
    "    new = oldSeq.replace(\"\\n\", \"\")\n",
    "    for number in range(0,10):\n",
    "        temp = new.replace(str(number), \"\")\n",
    "        new = temp\n",
    "    oldSeq = new\n",
    "    new = oldSeq.replace(\" \", \"\")\n",
    "    return new.upper()\n",
    "\n",
    "def appendPrimers(data, insert, tag, primerList, nameList):\n",
    "    name = str(data.ix[number,\"Gene\"]).replace(\" \", \"\") \n",
    "    nameFOR = name + \" FOR \" + tag + \"-tagged\"\n",
    "    nameREV = name + \" REV \" + tag + \"-tagged\"\n",
    "    primerList.append(screenPrimers(insert, leftbackbone , rightbackbone)[0])\n",
    "    nameList.append(nameFOR)\n",
    "    primerList.append(screenPrimers(insert, leftbackbone , rightbackbone)[1])\n",
    "    nameList.append(nameREV)\n",
    "\n",
    "\n",
    "primerList = []\n",
    "nameList = []\n",
    "\n",
    "for number in range(0,93):\n",
    "    tag = \"\"\n",
    "#We don't know whether unspecified proteins are N or C His-tagged, so we assume they could be either.\n",
    "    leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg'.upper())\n",
    "    rightbackbone = DNASequence('ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcgcgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgcca'.upper())\n",
    "    insert = DNASequence(formatSeq(data.ix[number,\"Sequence - DNA sequence\"]))\n",
    "    insert = DNASequence(insert.sequence[:len(insert.sequence) - 3])\n",
    "    if len(insert.sequence) != 0:\n",
    "        tag = \"C\"\n",
    "        appendPrimers(data, insert, tag, primerList, nameList)\n",
    "    leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatc'.upper())\n",
    "    rightbackbone = DNASequence('atggctgccgcgcggcaccaggccgctgctgtgatgatgatgatgatggctgctgcccatggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccag'.upper())\n",
    "    insert = DNASequence(formatSeq(data.ix[number,\"Sequence - DNA sequence\"]))\n",
    "    if len(insert.sequence) != 0:\n",
    "        tag = \"N\"\n",
    "        appendPrimers(data, insert, tag, primerList, nameList)\n",
    "        \n",
    "dataframe = pandas.DataFrame(primerList, nameList)\n",
    "\n",
    "writer = pandas.ExcelWriter('primerList2.xlsx', engine='openpyxl')\n",
    "\n",
    "dataframe.to_excel(writer)\n",
    "\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Sequencing Primer Program"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTCCGCTTCAATCACTTTACGTGCAG\n",
      "AGAGTACCTTCACGTACCAGAACGGTA\n",
      "CCAGAGGCCACTTTCGTTGAACG\n",
      "GATCGCTTTCGTCTTCTGCCTGG\n"
     ]
    }
   ],
   "source": [
    "import primer3\n",
    "import pandas\n",
    "import math\n",
    "\n",
    "\n",
    "def circularizePlasmid(plasmid):\n",
    "    half = int(len(plasmid.sequence) / 2)\n",
    "    #print(half)\n",
    "    plasmid2 = DNASequence(plasmid.sequence[half:] + plasmid.sequence[:half])\n",
    "    return (plasmid, plasmid2)\n",
    "\n",
    "def findSeqPrimers(startPrimer, endPrimer, plasmid): #make sure insert is in middle of sequence for now\n",
    "    plasmidToCheck = circularizePlasmid(plasmid)\n",
    "    seqPrimerArr = []\n",
    "    endPos = plasmidToCheck[0].sequence.find(reverseComplement(endPrimer.sequence)) - 500\n",
    "    \n",
    "    startPos = plasmidToCheck[0].sequence.find(startPrimer.sequence) + len(startPrimer.sequence)\n",
    "    currentPos = startPos + 500\n",
    "    while currentPos < endPos:\n",
    "        seqPrimer = createSeqPrimer(currentPos, plasmidToCheck[0], 18, 0)\n",
    "        seqPrimerArr.append(seqPrimer)\n",
    "        if seqPrimer == \"No primers found; please lower your search stringency\":\n",
    "            break\n",
    "        currentPos = plasmidToCheck[0].sequence.find(seqPrimer) + len(seqPrimer) + 500\n",
    "    \n",
    "    return seqPrimerArr\n",
    "        \n",
    "def createSeqPrimer(currentPos, plasmid, primerSize, recursionNumber):\n",
    "    #print(recursionNumber)\n",
    "    if recursionNumber > 950:\n",
    "        return \"No primers found; please lower your search stringency\"\n",
    "    elif primerSize < 40:\n",
    "        primer = DNASequence(plasmid.sequence[currentPos: currentPos + primerSize])\n",
    "    \n",
    "        isGood = True\n",
    "        if primer.Tm < 58.0 or primer.Tm > 59.5:\n",
    "            isGood = False\n",
    "        if float(primer.Homodimer.dg) < -7000.0:\n",
    "            isGood = False\n",
    "        if float(primer.Hairpin.dg) < -500.0:\n",
    "            isGood = False\n",
    "    \n",
    "            #return the shortest (18bp ≤) desireable primers\n",
    "        if isGood == True:\n",
    "            #screening returned a good primer; return the primer\n",
    "            return primer.sequence\n",
    "        else:\n",
    "            #primer failed screening, try again with a larger primer using recursion\n",
    "            return createSeqPrimer(currentPos, plasmid, primerSize + 1, recursionNumber + 1)\n",
    "    else:\n",
    "        primerSize = 18\n",
    "        return createSeqPrimer(currentPos - 1, plasmid, primerSize, recursionNumber + 1)\n",
    "\n",
    "forward = DNASequence('tccggatatagttcctcctttcagc'.upper())\n",
    "reverse = DNASequence('GGTGATGTCGGCGATATAGGC'.upper())\n",
    "plas = DNASequence('cgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgccaatccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatcttaAGCAATGGTACGTTGGATCTCGATGATTTCGAATACTTCGATCACATCGCCAGTGCGGACGTCGTTGTAGTTCTTAACGCCGATACCACATTCCATACCGTTACGGACTTCGTTAACGTCATCTTTGAAGCGGCGCAGGGACTCCAGCTCGCCTTCGTAGATAACCACGTTGTCACGCAGAACGCGGATCGGGTTGTGACGTTTAACCACACCTTCGGTAACCATACAGCCTGCGATGGCACCAAATTTCGGCGATTTGAACACGTCACGAACTTCCGCCAGACCGATAATCTGCTGTTTCAGTTCCGGAGACAGCATACCGCTCATCGCCGCTTTCACTTCGTCAATCAGGTTATAGATGACGGAGTAGTAACGCAGATCCAGGCTTTCCGCTTCAATCACTTTACGTGCAGAGGCATCAGCACGTACGTTAAAGCCAACCAGGATGGCGTTGGACGCCGCAGCCAGGGTGGCGTCGGTTTCGGTGATACCACCTACGCCAGAACCGATGATCTTCACTTTAACTTCGTCAGTAGACAGTTTCAGCAAGGAGTCGGAGATCGCTTCGACAGAACCCTGTACGTCTGCCTTCAGGACGATATTCACTTCGTGAACTTCGCCTTCGGTCATGTTGGCGAACATGTTCTCGAGTTTAGATTTCTGCTGACGCGCCAGTTTAACTTCGCGGAATTTACCCTGACGATAGAGTGCAACTTCACGCGCTTTCTTCTCGTCACGTACAACGGTAACTTCATCACCCGCAGCCGGTACGCCGGACAGGCCGAGGATTTCCACCGGAATGGACGGACCCGCTTCCAGCACTTCCTGACCCAGTTCGTTACGCATCGCACGAACACGACCGTATTCGAAGCCACACAGAACGATATCGCCCTTGTGCAGAGTACCTTCACGTACCAGAACGGTAGCAACCGGACCACGACCTTTATCGAGGAAGGATTCGATAACCGCACCGCTCGCCATACCTTTACGTACCGCTTTCAGCTCCAGAACTTCCGCCTGCAGCAGGATAGCGTCCAGCAGTTCATCGATACCGGTACCCGCTTTCGCAGATACGTGTACGAACTGGCTTTCACCGCCCCACTCTTCCGGCAGGATGCCGTACTGGGAGAGTTCGTTCTTAACGCGATCCGGATCAGCTTCTGGTTTATCGATCTTGTTCACTGCAACCACCACCGGTACCTGCGCCGCTTTCGCGTGCTGGATTGCTTCGATGGTCTGCGGCATCACACCGTCGTCGGCAGCAACAACCAGGACTACGATGTCCGTTGCCTGCGCACCACGAGCACGCATTGAAGTAAACGCGGCGTGCCCCGGGGTGTCCAGGAAGGTGATCATGCCGTTTTCAGTTTCAACGTGGTATGCACCAATGTGCTGGGTAATGCCGCCCGCTTCGCCAGAGGCCACTTTCGTTGAACGAATGTAGTCCAGCAGAGAGGTTTTACCGTGGTCAACGTGACCCATGATGGTCACAACCGGCGCGCGCGGTTCAGCCGCAGCACCCGTGTCACGGTCGCTCATTACCGCCTCTTCCAGCTCGTTTTCACGACGCAGGATAACTTTATGGCCCATCTCTTCAGCAACCAGCTGTGCGGTTTCCTGATCGATAACCTGGTTGATGGTTGCCATTGCGCCCAGTTTCATCATCGCTTTGATGACCTGAGAGCCTTTAACCGCCATCTTGTTCGCCAGTTCGCCAACGGTGATAGTTTCGCCGATCACAACGTCACGGTTAACGGCCTGAGCAGGCTTCTGGAAGCCTTGCTGCAGCGAAGAACCTTTACGTTTTCCGCCTTTACCGCCACGTACTGCTGCGCGTGCTTCTTCACGATCAGCTTTTGATTCAGCGTGTTTGTTGCCTTTCTTCGGACGCGCTGCTTTCGCGTTACGACCACGGCCACGGCCGCCTTCGACTTCACGATCGCTTTCGTCTTCTGCCTGGCGAGCATGTTGAGAAGTAGTGACGTGATAATCGCTGGAATCTTCAGTCGGTTCCGCGTTATCAGTCCATTTGTTTTCTTCCGCCATACGACGTGCTTCTTCAGCAACGCGACGTGCTTCTTCTTCGAGTTTACGACGCGCTTCTTCTTCAGCTTTACGCTTGAGCTCTGCAGCTTCCTGCTCACGGCGGGCTTTTTCAGCCTGGGCGTTTTTAGTCATATCGTCTTGTTGATTGCTCACTTTGTCTTTTTCCGCAGCTTCACGTTTCGCTTGTTCAGCAGCTTCACGCTTAGCTTGTTCTGCGGCCTCACGTTCAGCTTTTTGTTGCGCCTCGCGTTTAGCCGATTCTTCTGCCTCACGACGGGCTTGCTCTTCCGCTTCACGCTGCGCTTGCTCTTCCGCTGCAAGGCGTTCAGCCTCTTGCGGATCGCGTTTCACAAAGGTGCGTTTCTTGCGGACTTCGATTTGTACCGATTTGCTTTTTCCACCGGTACCAGGAATGTTAAGGGTGCTGCGTGTTTTACGTTGCAGCGTCAATTTGTCCGGGCCTGAATTTTTCTGATTCAGGTGGTCAATCAAAGTCTGTTTCTCTTGTGCAGACACAGAGTCGTCAGCAGACTTCCGGATACCTGCATCAGCAAATTGCTGTACCAGGCGTTCCACGGAGGTCTGTCGCTCTGCGGCCAGCGTTTTAATCGTTACATCTGTcatatggctgccgcgcggcaccaggccgctgctgtgatgatgatgatgatggctgctgcccatggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcg'.upper())\n",
    "\n",
    "beta = findSeqPrimers(forward, reverse, plas)\n",
    "for alpha in beta:\n",
    "    print(alpha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'count': 0, 'created_at': '2017-08-01T07:03:23.289667+00:00', 'description': 'Sequences generated from the primer program inputted into the vector are available here', 'id': 'lib_iP53B30G', 'modified_at': '2017-08-01T07:03:23.289667+00:00', 'name': 'Program Generated Vectors', 'owner': 'ent_GopUD1my', 'permissions': {'admin': True, 'appendable': True, 'owner': False, 'readable': True, 'writable': True}, 'sequences': [], 'type': 'INVENTORY'}\n"
     ]
    }
   ],
   "source": [
    "from benchlingapi import BenchlingAPI\n",
    "\n",
    "bench_api_key = 'sk_gnGGnlq0laAcbBcI9yVEIqpE7opJQ'\n",
    "benchlingapi = BenchlingAPI(bench_api_key)\n",
    "\n",
    "print(benchlingapi.find_folder('Program Generated Vectors', query='name', regex=True))\n",
    "\n",
    "an = {\"color\": \"#303030\", \"end\": 25, \"name\": \"Untitled\", \"start\": 16, \"strand\": 1, \"type\": \"gene\"}\n",
    "\n",
    "#print(benchlingapi.create_sequence('testSeqProgram', 'cagggcagccgg', True, 'lib_f0TfvHp8', description='this is a test sequence', annotations=[an]))\n",
    "\n",
    "\n",
    "import requests\n",
    "\n",
    "#resp = requests.get('https://api.benchling.com/v1/sequences', auth=(bench_api_key, ''))\n",
    "#if resp.status_code != 200:\n",
    "    # This means something went wrong.\n",
    "    #raise ApiError('GET /tasks/ {}'.format(resp.status_code))\n",
    "\n",
    "#print(resp.json()[0])\n",
    "# for todo_item in resp.json():\n",
    "#     print('{} {}'.format(todo_item['id'], todo_item['summary']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Response [201]>\n",
      "<Response [201]>\n",
      "<Response [201]>\n",
      "                                                      0\n",
      "seqFOR1 groL N-tagged              CAGCGCAACACCACCACCAG\n",
      "seqFOR2 groL N-tagged               TCAGCAGCGGTTTGCCTGC\n",
      "seqFOR1 Met C-tagged       AGCGGCTTTTACTTCTTCTTTAGAGGCT\n",
      "seqFOR2 Met C-tagged   CGAAATCTTCCAGGTTGAGATCGATATCATCA\n",
      "seqFOR3 Met C-tagged        CATTTCGCTGAAAGAGGGCAGATCAAA\n"
     ]
    }
   ],
   "source": [
    "#resp.json()['sequences'][0]\n",
    "\n",
    "seqPrimers = []\n",
    "seqPrimerNames = []\n",
    "\n",
    "def unpackPrimers(nameList, seqList, finNameList, finSeqList):\n",
    "    for names in nameList:\n",
    "        finNameList.append(names)\n",
    "    for primers in seqList:\n",
    "        finSeqList.append(primers)\n",
    "\n",
    "a = createVector(DNASequence('atgGCAGCTAAAGACGTAAAATTCGGTAACGACGCTCGTGTGAAAATGCTGCGCGGCGTAAACGTACTGGCAGATGCAGTGAAAGTTACCCTCGGTCCAAAAGGCCGTAACGTAGTTCTGGATAAATCTTTCGGTGCACCGACCATCACCAAAGATGGTGTTTCCGTTGCTCGTGAAATCGAACTGGAAGACAAGTTCGAAAATATGGGTGCGCAGATGGTGAAAGAAGTTGCCTCTAAAGCAAACGACGCTGCAGGCGACGGTACCACCACTGCAACCGTACTGGCTCAGGCTATCATCACTGAAGGTCTGAAAGCTGTTGCTGCGGGCATGAACCCGATGGACCTGAAACGTGGTATCGACAAAGCGGTTACCGCTGCAGTTGAAGAACTGAAAGCGCTGTCCGTACCATGCTCTGACTCTAAAGCGATTGCTCAGGTTGGTACCATCTCCGCTAACTCCGACGAAACCGTAGGTAAACTGATCGCTGAAGCGATGGACAAAGTCGGTAAAGAAGGCGTTATCACCGTTGAAGACGGTACCGGTCTGCAGGACGAACTGGACGTGGTTGAAGGTATGCAGTTCGACCGTGGCTACCTGTCTCCTTACTTCATCAACAAGCCGGAAACTGGCGCAGTAGAACTGGAAAGCCCGTTCATCCTGCTGGCTGACAAGAAAATCTCCAACATCCGCGAAATGCTGCCGGTTCTGGAAGCTGTTGCCAAAGCAGGCAAACCGCTGCTGATCATCGCTGAAGATGTAGAAGGCGAAGCGCTGGCAACTCTGGTTGTTAACACCATGCGTGGCATCGTGAAAGTCGCTGCGGTTAAAGCACCGGGCTTCGGCGATCGTCGTAAAGCTATGCTGCAGGATATCGCAACCCTGACTGGCGGTACCGTGATCTCTGAAGAGATCGGTATGGAGCTGGAAAAAGCAACCCTGGAAGACCTGGGTCAGGCTAAACGTGTTGTGATCAACAAAGACACCACCACTATCATCGATGGCGTGGGTGAAGAAGCTGCAATCCAGGGCCGTGTTGCTCAGATCCGTCAGCAGATTGAAGAAGCAACTTCTGACTACGACCGTGAAAAACTGCAGGAACGCGTAGCGAAACTGGCAGGCGGCGTTGCAGTTATCAAAGTGGGTGCTGCTACCGAAGTTGAAATGAAAGAGAAAAAAGCACGCGTTGAAGATGCCCTGCACGCGACCCGTGCTGCGGTAGAAGAAGGCGTGGTTGCTGGTGGTGGTGTTGCGCTGATCCGCGTAGCGTCTAAACTGGCTGACCTGCGTGGTCAGAACGAAGACCAGAACGTGGGTATCAAAGTTGCACTGCGTGCAATGGAAGCTCCGCTGCGTCAGATCGTATTGAACTGCGGCGAAGAACCGTCTGTTGTTGCTAACACCGTTAAAGGCGGCGACGGCAACTACGGTTACAACGCAGCAACCGAAGAATACGGCAACATGATCGACATGGGTATCCTGGATCCAACCAAAGTAACTCGTTCTGCTCTGCAGTACGCAGCTTCTGTGGCTGGCCTGATGATCACCACCGAATGCATGGTTACCGACCTGCCGAAAAACGATGCAGCTGACTTAGGCGCTGCTGGCGGTATGGGCGGCATGGGTGGCATGGGCGGCATGATGtaa'.upper()), 'N', \"groL\")\n",
    "b = createVector(DNASequence('atgAATATTCGTCCATTGCATGATCGCGTGATCGTCAAGCGTAAAGAAGTTGAAACTAAATCTGCTGGCGGCATCGTTCTGACCGGCTCTGCAGCGGCTAAATCCACCCGCGGCGAAGTGCTGGCTGTCGGCAATGGCCGTATCCTTGAAAATGGCGAAGTGAAGCCGCTGGATGTGAAAGTTGGCGACATCGTTATTTTCAACGATGGCTACGGTGTGAAATCTGAGAAGATCGACAATGAAGAAGTGTTGATCATGTCCGAAAGCGACATTCTGGCAATTGTTGAAGCGtaa'.upper()), 'N', \"groS\")\n",
    "c = createVector(DNASequence('atgACTCAAGTCGCGAAGAAAATTCTGGTGACGTGCGCACTGCCGTACGCTAACGGCTCAATCCACCTCGGCCATATGCTGGAGCACATCCAGGCTGATGTCTGGGTCCGTTACCAGCGAATGCGCGGCCACGAGGTCAACTTCATCTGCGCCGACGATGCCCACGGTACACCGATCATGCTGAAAGCTCAGCAGCTTGGTATCACCCCGGAGCAGATGATTGGCGAAATGAGTCAGGAGCATCAGACTGATTTCGCAGGCTTTAACATCAGCTATGACAACTATCACTCGACGCACAGCGAAGAGAACCGCCAGTTGTCAGAACTTATCTACTCTCGCCTGAAAGAAAACGGTTTTATTAAAAACCGCACCATCTCTCAGCTGTACGATCCGGAAAAAGGCATGTTCCTGCCGGACCGTTTTGTGAAAGGCACCTGCCCGAAATGTAAATCCCCGGATCAATACGGCGATAACTGCGAAGTCTGCGGCGCGACCTACAGCCCGACTGAACTGATCGAGCCGAAATCGGTGGTTTCTGGCGCTACGCCGGTAATGCGTGATTCTGAACACTTCTTCTTTGATCTGCCCTCTTTCAGCGAAATGTTGCAGGCATGGACCCGCAGCGGTGCGTTGCAGGAGCAGGTGGCAAATAAAATGCAGGAGTGGTTTGAATCTGGCCTGCAACAGTGGGATATCTCCCGCGACGCCCCTTACTTCGGTTTTGAAATTCCGAACGCGCCGGGCAAATATTTCTACGTCTGGCTGGACGCACCGATTGGCTACATGGGTTCTTTCAAGAATCTGTGCGACAAGCGCGGCGACAGCGTAAGCTTCGATGAATACTGGAAGAAAGACTCCACCGCCGAGCTGTACCACTTCATCGGTAAAGATATTGTTTACTTCCACAGCCTGTTCTGGCCTGCCATGCTGGAAGGCAGCAACTTCCGCAAGCCGTCCAACCTGTTTGTTCATGGCTATGTGACGGTGAACGGCGCAAAGATGTCCAAGTCTCGCGGCACCTTTATTAAAGCCAGCACCTGGCTGAATCATTTTGACGCAGACAGCCTGCGTTACTACTACACTGCGAAACTCTCTTCGCGCATTGATGATATCGATCTCAACCTGGAAGATTTCGTTCAGCGTGTGAATGCCGATATCGTTAACAAAGTGGTTAACCTGGCCTCCCGTAATGCGGGCTTTATCAACAAGCGTTTTGACGGCGTGCTGGCAAGCGAACTGGCTGACCCGCAGTTGTACAAAACCTTCACTGATGCCGCTGAAGTGATTGGTGAAGCGTGGGAAAGCCGTGAATTTGGTAAAGCCGTGCGCGAAATCATGGCGCTGGCTGATCTGGCTAACCGCTATGTCGATGAACAGGCTCCGTGGGTGGTGGCGAAACAGGAAGGCCGCGATGCCGACCTGCAGGCAATTTGCTCAATGGGCATCAACCTGTTCCGCGTGCTGATGACTTACCTGAAGCCGGTACTGCCGAAACTGACCGAGCGTGCAGAAGCATTCCTCAATACGGAACTGACCTGGGATGGTATCCAGCAACCGCTGCTGGGCCACAAAGTGAATCCGTTCAAGGCGCTGTATAACCGCATCGATATGAGGCAGGTTGAAGCACTGGTGGAAGCCTCTAAAGAAGAAGTAAAAGCCGCTGCCGCGCCGGTAACTGGCCCGCTGGCAGATGATCCGATTCAGGAAACCATCACCTTTGACGACTTCGCTAAAGTTGACCTGCGCGTGGCGCTGATTGAAAACGCAGAGTTTGTTGAAGGTTCTGACAAACTGCTGCGCCTGACGCTGGATCTCGGCGGTGAAAAACGCAATGTCTTCTCCGGTATTCGTTCTGCTTACCCGGATCCGCAGGCACTGATTGGTCGTCACACCATTATGGTGGCTAACCTGGCACCACGTAAAATGCGCTTCGGTATCTCTGAAGGCATGGTGATGGCTGCCGGTCCTGGCGGGAAAGATATTTTCCTGCTAAGCCCGGATGCCGGTGCTAAACCGGGTCATCAGGTGAAAtaa'.upper()), \"C\", \"Met\")\n",
    "                 \n",
    "unpackPrimers(a[0], a[1], seqPrimerNames, seqPrimers)\n",
    "unpackPrimers(b[0], b[1], seqPrimerNames, seqPrimers)\n",
    "unpackPrimers(c[0], c[1], seqPrimerNames, seqPrimers)\n",
    "\n",
    "df = pd.DataFrame(seqPrimers, seqPrimerNames)\n",
    "print(df)\n",
    "\n",
    "writer = pd.ExcelWriter('test3.xlsx', engine='openpyxl')\n",
    "\n",
    "df.to_excel(writer)\n",
    "\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def createVector(insert, tag, name):\n",
    "    seqPrimerList = []\n",
    "    seqPrimerNameList = []\n",
    "    insert = insert.sequence\n",
    "    annotations = []\n",
    "    pBR322_origin = \"cgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatct\".upper()\n",
    "    kanR2 = \"atgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaa\".upper()\n",
    "    rop = \"gtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctga\".upper()\n",
    "    t7Terminator = \"atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcag\".upper()\n",
    "    tet300 = \"gccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagag\".upper()\n",
    "    pGEX_3_primer = \"cctctgacacatgcagctcccgg\".upper()\n",
    "    thrombin = thrombin = \"gctgccgcgcggcaccag\".upper()\n",
    "    t7promoter = \"cctatagtgagtcgtatta\".upper()\n",
    "    lacO = \"ggggaattgttatccgctcacaattccc\".upper()\n",
    "    hisTag = \"gtgatgatgatgatgatg\".upper()\n",
    "    seqRev = \"gcctatatcgccgacatcacc\".upper()\n",
    "    seqFor = \"tccggatatagttcctcctttcagc\".upper()\n",
    "    rbs = \"ctccttcttaaagttaa\".upper()\n",
    "    hisPlusStop = \"tcagtggtggtggtggtggtg\".upper()\n",
    "    lacI = \"gtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtga\".upper()\n",
    "    tet576 = \"cccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatg\".upper()\n",
    "    \n",
    "    \n",
    "    if tag ==\"C\":\n",
    "        leftBackbone = \"cgcgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgccaatccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg\".upper()\n",
    "        rightBackbone = \"ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcaggg\".upper()\n",
    "        sequence = leftBackbone + reverseComplement(insert[:len(insert) - 3]) + rightBackbone\n",
    "        endLen = len(leftBackbone) + len(insert) - 3\n",
    "        addAnnotation(annotations, reverseComplement(insert[:len(insert) - 3]), sequence, name, 'gene', \"#0000ff\")\n",
    "        \n",
    "    if tag ==\"N\":\n",
    "        leftBackbone = \"cgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgccaatccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatc\".upper()\n",
    "        rightBackbone = \"atggctgccgcgcggcaccaggccgctgctgtgatgatgatgatgatggctgctgcccatggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcg\".upper()\n",
    "        sequence = leftBackbone + reverseComplement(insert) + rightBackbone\n",
    "        endLen = len(leftBackbone) + len(insert)\n",
    "        addAnnotation(annotations, reverseComplement(insert), sequence, name, 'gene', \"#0000ff\")\n",
    "        \n",
    "    offName = str(name) + \" \" + tag + \"-tagged\"\n",
    "    \n",
    "    beta = findSeqPrimers(DNASequence(seqFor), DNASequence(reverseComplement(seqRev)), DNASequence(sequence))\n",
    "    count = 0\n",
    "    for alpha in beta:\n",
    "        count += 1\n",
    "        nom = \"seqFOR\" + str(count) + \" \" + offName\n",
    "        addAnnotation(annotations, alpha, sequence, nom, \"primer\", \"#f2c22e\")\n",
    "        seqPrimerNameList.append(nom)\n",
    "        seqPrimerList.append(alpha)\n",
    "    \n",
    "    addAnnotation(annotations, pBR322_origin, sequence, \"pBR322 Origin\", \"origin\", \"#9faa9a\")\n",
    "    addAnnotation(annotations, kanR2, sequence, \"Kan R2\", \"gene\", \"#359797\")\n",
    "    addAnnotation(annotations, t7Terminator, sequence, \"T7 Terminator\", \"terminator\", \"#d24b19\")\n",
    "    addAnnotation(annotations, tet300, sequence, \"tet (300-563)\", \"gene\", \"#ffe2b9\")\n",
    "    addAnnotation(annotations, pGEX_3_primer, sequence, \"pGEX_3_primer\", \"primer\", \"#5b637e\")\n",
    "    addAnnotation(annotations, thrombin, sequence, \"thrombin\", \"tag\", \"#5f631d\")\n",
    "    addAnnotation(annotations, t7promoter, sequence, \"T7 Promoter\", \"promoter\", \"#c8498b\")\n",
    "    addAnnotation(annotations, lacO, sequence, \"lacO\", \"regulatory\", \"#fffbaa\")\n",
    "    addAnnotation(annotations, lacO, sequence, \"lacI\", \"regulatory\", \"#c3dff3\")\n",
    "    addAnnotation(annotations, hisTag, sequence, \"His Tag\", \"tag\", \"#f9c5fe\")\n",
    "    addAnnotation(annotations, seqFor, sequence, \"pET28+ Forward Primer\", \"primer\", \"#fffafa\")\n",
    "    addAnnotation(annotations, seqRev, sequence, \"pET28+ Reverse Primer\", \"primer\", \"#fffafa\")\n",
    "    addAnnotation(annotations, rbs, sequence, \"T7 Ribosome Binding Site\", \"feature\", \"#e12222\")\n",
    "    addAnnotation(annotations, hisPlusStop, sequence, \"His Tag and Stop Codon\", \"feature\", \"#ddb34e\")\n",
    "    addAnnotation(annotations, tet576, sequence, \"tet(576-851)\", \"feature\", \"#ba9231\")\n",
    "    addAnnotation(annotations, rop, sequence, \"ROP\", \"feature\", \"#1ebdac\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    data = {\n",
    "      \"aliases\": [],\n",
    "      \"annotations\": annotations,\n",
    "      \"bases\": sequence,\n",
    "      \"circular\": True,\n",
    "      \"description\": \"program generated\",\n",
    "      \"folder\": \"lib_iP53B30G\",\n",
    "      \"name\": offName,\n",
    "      \"tags\": []\n",
    "    }\n",
    "\n",
    "    key = 'sk_gnGGnlq0laAcbBcI9yVEIqpE7opJQ'\n",
    "    resp = requests.post('https://api.benchling.com/v1/sequences/', json=data, auth=(key, ''))\n",
    "    print(resp)\n",
    "    return (seqPrimerNameList, seqPrimerList)\n",
    "\n",
    "def findSequence(sequence, plasmid):\n",
    "    if sequence in plasmid:\n",
    "        a = plasmid.find(sequence)\n",
    "        #print(a, a + len(sequence))\n",
    "    else:\n",
    "        return (-1, -1)\n",
    "    return (a, a + len(sequence))\n",
    "def addAnnotation(annotations, sequence, plasmid, name, annotationType, color):\n",
    "    annotation = {\n",
    "          \"color\": color,\n",
    "          \"end\": findSequence(sequence, plasmid)[1],\n",
    "          \"name\": name,\n",
    "          \"start\": findSequence(sequence, plasmid)[0],\n",
    "          \"strand\": 0,\n",
    "          \"type\": annotationType\n",
    "    }\n",
    "    if findSequence(sequence, plasmid) != (-1, -1):\n",
    "        annotations.append(annotation)\n",
    "    return annotations\n",
    "    \n",
    "#createVector('ATGGAAAAGACATATAACCCACAAGATATCGAACAGCCGCTTTACGAGCACTGGGAAAAGCAGGGCTACTTTAAGCCTAATGGCGATGAAAGCCAGGAAAGTTTCTGCATCATGATCCCGCCGCCGAACGTCACCGGCAGTTTGCATATGGGTCACGCCTTCCAGCAAACCATCATGGATACCATGATCCGCTATCAGCGCATGCAGGGCAAAAACACCCTGTGGCAGGTCGGTACTGACCACGCCGGGATCGCTACCCAGATGGTCGTTGAGCGCAAGATTGCCGCAGAAGAAGGTAAAACCCGTCACGACTACGGCCGCGAAGCTTTCATCGACAAAATCTGGGAATGGAAAGCGGAATCTGGCGGCACCATTACCCGTCAGATGCGCCGTCTCGGCAACTCCGTCGACTGGGAGCGTGAACGCTTCACCATGGACGAAGGCCTGTCCAATGCGGTGAAAGAAGTTTTCGTTCGTCTGTATAAAGAAGACCTGATTTACCGTGGCAAACGCCTGGTAAACTGGGATCCGAAACTGCGCACCGCTATCTCTGACCTGGAAGTCGAAAACCGCGAATCGAAAGGTTCGATGTGGCACATCCGCTATCCGCTGGCTGACGGTGCGAAAACCGCAGACGGTAAAGATTATCTGGTGGTCGCGACTACCCGTCCAGAAACCCTGCTGGGCGATACTGGCGTAGCCGTTAACCCGGAAGATCCGCGTTACAAAGATCTGATTGGCAAATATGTCATTCTGCCGCTGGTTAACCGTCGTATTCCGATCGTTGGCGACGAACACGCCGACATGGAAAAAGGCACCGGCTGCGTGAAAATCACTCCGGCGCACGACTTTAACGACTATGAAGTGGGTAAACGTCACGCCCTGCCGATGATCAACATCCTGACCTTTGACGGCGATATCCGTGAAAGCGCCCAGGTGTTCGATACCAAAGGTAACGAATCTGACGTTTATTCCAGCGAAATCCCTGCAGAGTTCCAGAAACTGGAGCGTTTTGCTGCACGTAAAGCAGTCGTTGCCGCAGTTGACGCGCTTGGCCTGCTGGAAGAAATTAAACCGCACGACCTGACCGTTCCTTACGGCGACCGTGGCGGCGTAGTTATCGAACCAATGCTGACCGACCAGTGGTACGTGCGTGCCGATGTCCTGGCGAAACCGGCGGTTGAAGCGGTTGAGAACGGCGACATTCAGTTCGTACCGAAGCAGTACGAAAACATGTACTTCTCCTGGATGCGCGATATTCAGGACTGGTGTATCTCTCGTCAGTTGTGGTGGGGTCACCGTATCCCGGCATGGTATGACGAAGCGGGTAACGTTTATGTTGGCCGCAACGAAGACGAAGTGCGTAAAGAAAATAACCTCGGTGCTGATGTTGTCCTGCGTCAGGACGAAGACGTTCTCGATACCTGGTTCTCTTCTGCGCTGTGGACCTTCTCTACCCTTGGCTGGCCGGAAAATACCGACGCCCTGCGTCAGTTCCACCCAACCAGCGTGATGGTATCTGGTTTCGACATCATTTTCTTCTGGATTGCCCGCATGATCATGATGACCATGCACTTCATCAAAGATGAAAATGGCAAACCGCAGGTGCCGTTCCACACCGTTTACATGACCGGCCTGATTCGTGATGACGAAGGCCAGAAGATGTCCAAATCCAAGGGTAACGTTATCGACCCACTGGATATGGTTGACGGTATTTCGCTGCCAGAACTGCTGGAAAAACGTACCGGCAATATGATGCAGCCGCAGCTGGCGGACAAAATCCGTAAGCGCACCGAGAAGCAGTTCCCGAACGGTATTGAGCCGCACGGTACTGACGCGCTGCGCTTCACCCTGGCGGCGCTGGCGTCTACCGGTCGTGACATCAACTGGGATATGAAGCGTCTGGAAGGTTACCGTAACTTCTGTAACAAGCTGTGGAACGCCAGCCGCTTTGTGCTGATGAACACAGAAGGTCAGGATTGCGGCTTCAACGGCGGCGAAATGACGCTGTCGCTGGCGGACCGCTGGATTCTGGCGGAGTTCAACCAGACCATCAAAGCGTACCGCGAAGCGCTGGACAGCTTCCGCTTCGATATCGCCGCAGGCATTCTGTATGAGTTCACCTGGAACCAGTTCTGTGACTGGTATCTCGAGCTGACCAAGCCGGTAATGAACGGTGGCACCGAAGCAGAACTGCGCGGTACTCGCCATACGCTGGTGACTGTACTGGAAGGTCTGCTGCGCCTCGCGCATCCGATCATTCCGTTCATCACCGAAACCATCTGGCAGCGTGTGAAAGTACTTTGCGGTATCACTGCCGACACCATCATGCTGCAGCCGTTCCCGCAGTACGATGCATCTCAGGTTGATGAAGCCGCACTGGCCGACACCGAATGGCTGAAACAGGCGATCGTTGCGGTACGTAACATCCGTGCAGAAATGAACATCGCGCCGGGCAAACCGCTGGAGCTGCTGCTGCGTGGTTGCAGCGCGGATGCAGAACGTCGCGTAAATGAAAACCGTGGCTTCCTGCAAACCCTGGCGCGTCTGGAAAGTATCACCGTGCTGCCTGCCGATGACAAAGGTCCGGTTTCCGTTACGAAGATCATCGACGGTGCAGAGCTGCTGATCCCGATGGCTGGCCTCATCAACAAAGAAGATGAGCTGGCGCGTCTGGCGAAAGAAGTGGCGAAGATTGAAGGTGAAATCAGCCGTATCGAGAACAAACTGGCGAACGAAGGCTTTGTCGCCCGCGCACCGGAAGCGGTCATCGCGAAAGAGCGTGAGAAGCTGGAAGGCTATGCGGAAGCGAAAGCGAAACTGATTGAACAGCAGGCTGTTATCGCCGCGCTGtaa'.upper(), 'C')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
