{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def findHomology(sequence1, sequence2): #Returns largest homologous region between the 2 sequences\n",
    "    m = [[0] * (1+len(sequence2))\n",
    "    for i in range(1+len(sequence1))]\n",
    "        longest, x_longest = 0, 0\n",
    "        for x in range(1, 1 + len(sequence2)):\n",
    "            for y in range(1, 1 + len(sequence2)):\n",
    "                if sequence1[x - 1] == sequence2[y - 1]:\n",
    "                    m[x][y] = m[x - 1][y - 1] + 1\n",
    "                    if m[x][y] > longest:\n",
    "                        longest = m[x][y]\n",
    "                        x_longest = x\n",
    "                else:\n",
    "                    m[x][y] = 0\n",
    "    return sequence1[x_longest - longest: x_longest]\n",
    "        \n",
    "def choosePrimers(insert, upstreamBackbone, downstreamBackbone): #Make primers for an insert to amplify it such that it can fit in a \n",
    "    forwardPrimer = insert[:20]\n",
    "    reversePrimer = reverseComplement(insert[len(insert) - 20: len(insert)])\n",
    "    getTm\n",
    "    \n",
    "def reverseComplement(sequence):\n",
    "    reverseComplement = \"\"\n",
    "    for base in sequence:\n",
    "        if base == \"a\":\n",
    "            reverseComplement = reverseComplement + \"t\"\n",
    "        elif base == \"t\":\n",
    "            reverseComplement = reverseComplement + \"a\"\n",
    "        elif base == \"g\":\n",
    "            reverseComplement = reverseComplement + \"c\"\n",
    "        elif base == \"c\":\n",
    "            reverseComplement = reverseComplement + \"g\"\n",
    "        else:\n",
    "            print (\"error\")\n",
    "    return (reverseComplement)\n",
    "def getTm(sequence):\n",
    "    composition = getComposititon(sequence)\n",
    "    #Tm= 64.9 +41*(yG+zC-16.4)/(wA+xT+yG+zC) \n",
    "    Tm = 64.9 + 41*(composition[\"G\"] + composition[\"C\"] - 16.4) / (composition[\"A\"] + composition[\"T\"] + composition[\"G\"] + composition[\"C\"])\n",
    "    return (Tm)\n",
    "def getComposition(sequence):\n",
    "    composition = {\"A\" : 0, \"G\" : 0, \"C\" : 0, \"T\" : 0}\n",
    "    for base in sequence:\n",
    "        if base == \"a\":\n",
    "            composition[\"A\"] = composition[\"A\"] + 1\n",
    "        elif base == \"t\":\n",
    "            composition[\"T\"] = composition[\"T\"] + 1\n",
    "        elif base == \"g\":\n",
    "            composition[\"G\"] = composition[\"G\"] + 1\n",
    "        elif base == \"c\":\n",
    "            composition[\"C\"] = composition[\"C\"] + 1\n",
    "        else:\n",
    "            print(\"error 2\")\n",
    "    return composititon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGGAAAAGACATATAACCCACAAGATATCGAACA forward\n",
      "CAGCGCGGCGATAACAGCC reverse\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "('CTAGAAATAATTTTGTTTAACTTTAAGAAGGAGATATACCATGGAAAAGACATATAACCCACAAGATATCGAACA',\n",
       " 'CTTTGTTAGCAGCCGGATCTCAGTGGTGGTGGTGGTGGTGCAGCGCGGCGATAACAGCC')"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import primer3\n",
    "\n",
    "class DNASequence():\n",
    "    def __init__(self, sequence):\n",
    "        #Initialize the DNA variable, which has sequence info as well as additional properties (Tm, homodimer & hairpin formation)\n",
    "        self.sequence = sequence\n",
    "        self.composition = self.getComposition(sequence)\n",
    "        if len(sequence) != 0:\n",
    "            self.gcContent = self.getGCcontent(sequence)\n",
    "        else:\n",
    "            self.gcContent = 0\n",
    "        if len(sequence) < 60:\n",
    "            self.Tm = primer3.calcTm(sequence)\n",
    "            self.Homodimer = primer3.calcHomodimer(sequence)\n",
    "            self.Hairpin = primer3.calcHairpin(sequence)\n",
    "        else:\n",
    "            self.Tm = 100000 #set the Tm arbitrarily high\n",
    "        \n",
    "    def getComposition(self, sequence):\n",
    "        composition = {\"A\" : 0, \"G\" : 0, \"C\" : 0, \"T\" : 0}\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                composition[\"A\"] = composition[\"A\"] + 1\n",
    "            elif base == \"T\":\n",
    "                composition[\"T\"] = composition[\"T\"] + 1\n",
    "            elif base == \"G\":\n",
    "                composition[\"G\"] = composition[\"G\"] + 1\n",
    "            elif base == \"C\":\n",
    "                composition[\"C\"] = composition[\"C\"] + 1\n",
    "            else:\n",
    "                print(\"error 2\")\n",
    "        return composition\n",
    "    def getGCcontent(self, sequence):\n",
    "        GCpercent = (self.composition[\"G\"] + self.composition[\"C\"])/(self.composition[\"G\"] + self.composition[\"T\"] + self.composition[\"A\"] + self.composition[\"C\"])\n",
    "        return GCpercent\n",
    "    def reverseComplement(self, sequence):\n",
    "        #Takes the complement of the sequence and reverses it.\n",
    "        reverseComplement = \"\"\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                reverseComplement = \"T\" + reverseComplement\n",
    "            elif base == \"T\":\n",
    "                reverseComplement = \"A\" + reverseComplement\n",
    "            elif base == \"G\":\n",
    "                reverseComplement = \"C\" + reverseComplement\n",
    "            elif base == \"C\":\n",
    "                reverseComplement = \"G\" + reverseComplement \n",
    "            else:\n",
    "                print (\"error\")\n",
    "        return (reverseComplement)\n",
    "def reverseComplement(sequence):\n",
    "    #Takes the complement of the sequence and reverses it.\n",
    "    reverseComplement = \"\"\n",
    "    for base in sequence:\n",
    "        if base == \"A\":\n",
    "            reverseComplement = \"T\" + reverseComplement \n",
    "        elif base == \"T\":\n",
    "            reverseComplement = \"A\" + reverseComplement\n",
    "        elif base == \"G\":\n",
    "            reverseComplement = \"C\" + reverseComplement\n",
    "        elif base == \"C\":\n",
    "            reverseComplement = \"G\" + reverseComplement\n",
    "        else:\n",
    "            print (\"error\")\n",
    "    return (reverseComplement)\n",
    "def screenPrimers(insert, upstreamBackbone, downstreamBackbone):\n",
    "    #run the main program to create and screen the forward and reverse primers and then add overhangs.\n",
    "    forwardPrimer = screenPrimer(insert, \"F\")\n",
    "    print(forwardPrimer.sequence, \"forward\")\n",
    "    reversePrimer = screenPrimer(insert, \"R\")\n",
    "    print(reversePrimer.sequence, \"reverse\")\n",
    "    \n",
    "    primers = addOverhangs(forwardPrimer, reversePrimer, \"reverse\", upstreamBackbone, downstreamBackbone)\n",
    "    return (primers[0].sequence, primers[1].sequence)\n",
    "def screenPrimer(insert, whichPrimer, primerSize=18): #Make primers for an insert to amplify it such that it can be assembled with Gibson\n",
    "    if primerSize < 40:\n",
    "        #Make sure you are selecting the right region to prime. \n",
    "        #This algorithm creates primers that amplify the WHOLE SEQUENCE minus the stop codon at the end.\n",
    "        if whichPrimer == \"F\": \n",
    "            primer = DNASequence(insert.sequence[:primerSize])\n",
    "        elif whichPrimer == \"R\":\n",
    "            primer = DNASequence(reverseComplement(insert.sequence[len(insert.sequence) - primerSize: len(insert.sequence)]))\n",
    "        else:\n",
    "            print(\"error\")\n",
    "        #Add various screens below here. You can update the parameters to be more or less stringent depending on how likely you would like your primers to work\n",
    "        isGood = True\n",
    "        if primer.Tm < 58.0 or primer.Tm > 59.5:\n",
    "            isGood = False\n",
    "        if float(primer.Homodimer.dg) < -10000.0:\n",
    "            isGood = False\n",
    "        if float(primer.Hairpin.dg) < -5000.0:\n",
    "            isGood = False\n",
    "    \n",
    "        #return the shortest (18bp â‰¤) desireable primers\n",
    "        if isGood == True:\n",
    "            #screening returned a good primer; return the primer\n",
    "            return primer\n",
    "        else:\n",
    "            #primer failed screening, try again with a larger primer using recursion\n",
    "            if whichPrimer == \"F\":\n",
    "                primer = screenPrimer(insert, \"F\", primerSize + 1) \n",
    "            elif whichPrimer == \"R\":\n",
    "                primer = screenPrimer(insert, \"R\", primerSize + 1)\n",
    "            return primer\n",
    "    else:\n",
    "        #parameters are too stringent; no primers were found that matched the criteria given. Return an empty string\n",
    "        print (\"You cannot make primers with the given parameters\")\n",
    "        return DNASequence(\"\")\n",
    "def addOverhangs(primer1, primer2, orientation, leftBackbone, rightBackbone):\n",
    "    #decide how the construct should be oriented in the final vector and add the overhangs so Gibson will add correctly.\n",
    "    if orientation == \"forward\":\n",
    "        primer2.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer2.sequence\n",
    "        primer1.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer1.sequence\n",
    "        return primer1, primer2\n",
    "    elif orientation == \"reverse\":\n",
    "        primer1.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer1.sequence\n",
    "        primer2.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer2.sequence\n",
    "        return primer1, primer2\n",
    "    else:\n",
    "        print (\"please enter either a forward or reverse strand orientation\")\n",
    "\n",
    "\n",
    "        \n",
    "insert = DNASequence('atgGAAAAGACATATAACCCACAAGATATCGAACAGCCGCTTTACGAGCACTGGGAAAAGCTGATTGAACAGCAGGCTGTTATCGCCGCGCTG'.upper())\n",
    "leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg'.upper())\n",
    "\n",
    "rightbackbone = DNASequence('ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcgcgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgcca'.upper())\n",
    "\n",
    "screenPrimers(insert, leftbackbone , rightbackbone)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACGCATTGAAACGGTCTTCAACG'"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reverseComplement('CGTTGAAGACCGTTTCAATGCGT'.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTAGAAATAATTTTGTTTAACTTTAAGAAGGAGATATACC\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CTTTGTTAGCAGCCGGATCTCAGTGGTGGTGGTGGTGGTG'"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import primer3\n",
    "\n",
    "primer3.calcTm('atggaaaagacatataacccacaagatatcgaac'.upper())\n",
    "\n",
    "t = (reverseComplement('ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgc'[:40].upper()))\n",
    "\n",
    "print(t)\n",
    "\n",
    "leftbackbone = ('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg'.upper())\n",
    "\n",
    "def reverse(sequence):\n",
    "    reverse = \"\"\n",
    "    for base in sequence:\n",
    "        if base == \"A\":\n",
    "            reverse = \"A\" + reverse \n",
    "        elif base == \"T\":\n",
    "            reverse = \"T\" + reverse\n",
    "        elif base == \"G\":\n",
    "            reverse = \"G\" + reverse\n",
    "        elif base == \"C\":\n",
    "            reverse = \"C\" + reverse\n",
    "        else:\n",
    "            print (\"error\")\n",
    "    return (reverse)\n",
    "\n",
    "s = (leftbackbone)\n",
    "\n",
    "s[len(s) - 40: len(s)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CGTTGAAGACCGTTTCAAT\n"
     ]
    }
   ],
   "source": [
    "def long_substr(data):\n",
    "    substr = ''\n",
    "    if len(data) > 1 and len(data[0]) > 0:\n",
    "        for i in range(len(data[0])):\n",
    "            for j in range(len(data[0])-i+1):\n",
    "                if j > len(substr) and is_substr(data[0][i:i+j], data):\n",
    "                    substr = data[0][i:i+j]\n",
    "    return substr\n",
    "\n",
    "def is_substr(find, data):\n",
    "    if len(data) < 1 and len(find) < 1:\n",
    "        return False\n",
    "    for i in range(len(data)):\n",
    "        if find not in data[i]:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "\n",
    "print (long_substr(['TCCCTATCAGTGATAGAGAAAAGAGAATCGGATCCAAAAGATCTAAAGAGGAGAAAGGATCTATGTTAATTGGCTATGTCCGTGTGTCCACCAACGATCAAAACACAGATCTTCAAAGAAATGCCCTCGTTTGCGCGGGATGTGAGCAGATCTTTGAAGATAAACTTAGCGGTACCCGTACCGATCGACCGGGTCTTAAACGCGCACTTAAACGGCTTCAGAAGGGGGATACCTTGGTCGTATGGAAACTGGACCGGCTGGGTAGATCGATGAAACATCTGATTAGTTTAGTTGGAGAACTGCGAGAACGCGGCATCAATTTTCGCTCCTTAACCGATAGCATCGATACCTCCAGTCCAATGGGGAGGTTTTTCTTCTATGTTATGGGCGCGCTGGCCGAAATGGAACGTGAATTGATTATCGAGCGGACAATGGCTGGGTTAGCGGCAGCACGAAATAAAGGACGGCGCTTCGGTCGTCCACCCAAGGGCGGCTCCGGTGGCTCGGGAGGTTCCGGCGGTAGCGGGGGCAGCGGCGGCAGTGGGGGCTCAGGAGGTTCTGATAAAAAATATTCTATAGGACTGGCTATTGGTACCAATTCAGTGGGTTGGGCGGTAATTACCGACGAATATAAAGTGCCGTCTAAAAAATTTAAGGTCTTGGGGAATACAGACCGTCACAGTATAAAAAAAAATCTGATCGGTGCGCTGCTGTTTGACTCTGGAGAAACTGCCGAAGCAACTCGTTTAAAACGAACCGCACGCCGTCGCTATACAAGGCGGAAAAATCGTATATGCTATTTACAAGAAATTTTTTCCAATGAAATGGCAAAAGTCGATGACAGCTTCTTTCATCGCTTAGAAGAGTCCTTTTTAGTTGAAGAAGATAAAAAACACGAACGCCATCCAATTTTTGGGAATATAGTTGATGAAGTCGCCTATCACGAAAAATATCCGACTATTTACCACCTGCGTAAGAAACTTGTGGATTCGACGGATAAAGCAGATCTGAGGCTGATTTATCTCGCCCTCGCACATATGATTAAATTCAGAGGTCATTTCCTGATAGAAGGCGATCTGAATCCCGACAACTCTGATGTTGATAAGTTGTTTATACAGCTGGTGCAAACCTATAATCAACTATTCGAAGAGAATCCTATAAACGCCAGCGGAGTGGACGCTAAAGCGATACTCAGCGCGCGCTTATCTAAAAGCCGGCGGCTGGAAAATCTGATCGCACAATTGCCAGGTGAAAAAAAAAATGGTTTATTTGGCAATCTGATAGCGCTTTCACTGGGGCTAACACCCAATTTTAAATCGAATTTTGATTTAGCCGAAGATGCTAAGCTGCAGCTGAGTAAAGATACCTATGATGACGATCTTGATAACCTGCTGGCACAAATTGGAGATCAATACGCAGATCTGTTTTTGGCCGCCAAGAATCTGTCGGATGCAATTCTGCTGAGCGATATTTTGAGGGTGAATACTGAAATCACTAAGGCACCGCTCTCTGCAAGTATGATCAAGCGTTATGATGAGCATCACCAGGACTTAACCCTGTTAAAAGCTCTGGTCCGCCAGCAGCTCCCGGAAAAATACAAGGAAATATTCTTCGATCAATCTAAGAATGGGTATGCCGGCTATATCGATGGCGGAGCGTCTCAGGAAGAATTTTATAAATTCATAAAACCTATCCTTGAAAAAATGGATGGAACGGAAGAGCTGCTGGTTAAACTGAATAGAGAGGATCTGTTACGCAAGCAGAGGACATTCGATAACGGTAGTATCCCACATCAAATACATCTGGGTGAACTTCACGCGATTCTTCGCCGCCAGGAAGACTTTTATCCCTTTCTAAAAGACAACAGGGAGAAAATTGAAAAAATCCTGACGTTTCGTATTCCTTATTATGTGGGGCCGCTGGCTCGTGGAAACAGTCGGTTCGCATGGATGACGCGCAAGTCCGAAGAGACGATAACGCCGTGGAATTTCGAAGAAGTGGTCGACAAAGGCGCAAGTGCACAGTCATTCATCGAACGTATGACTAACTTCGATAAAAATTTACCAAACGAGAAAGTTCTGCCTAAGCATAGCCTTCTCTATGAATACTTTACAGTCTATAACGAGTTAACTAAAGTTAAATATGTGACTGAAGGTATGCGTAAACCGGCGTTTCTGAGCGGCGAACAAAAAAAGGCAATTGTCGATCTGTTATTTAAAACAAATCGTAAAGTTACGGTCAAGCAATTGAAAGAGGACTACTTCAAAAAAATTGAGTGTTTCGACAGCGTCGAAATTTCTGGCGTTGAAGACCGTTTCAAT', 'CGTTGAAGACCGTTTCAATGCGTCGCTGGGAACTTACCACGATCTACTCAAAATCATAAAAGATAAAGACTTTCTGGACAATGAAGAGAACGAAGACATACTTGAAGACATCGTTCTGACCCTTACGCTATTTGAAGATAGGGAGATGATCGAAGAGCGCTTGAAGACCTATGCGCACCTATTCGATGATAAAGTTATGAAACAGCTCAAGCGGCGCCGTTACACGGGGTGGGGGCGTCTGTCGAGGAAGTTAATCAATGGAATTCGTGATAAGCAGTCAGGCAAGACGATCCTGGACTTTCTGAAATCAGATGGGTTTGCAAATCGTAACTTCATGCAGCTGATCCATGATGATTCGTTAACTTTTAAAGAAGATATTCAAAAAGCCCAAGTATCTGGTCAGGGTGACTCTCTTCACGAACATATCGCGAATTTGGCAGGTTCACCAGCTATTAAAAAAGGTATCTTGCAGACTGTGAAAGTTGTCGATGAACTTGTTAAAGTTATGGGTCGTCACAAGCCAGAGAATATCGTAATTGAGATGGCGCGTGAAAATCAAACAACCCAAAAAGGCCAGAAAAACTCGAGGGAACGCATGAAACGTATTGAAGAAGGTATTAAGGAGTTGGGTTCCCAGATACTTAAAGAACACCCGGTCGAAAATACGCAACTGCAAAATGAAAAATTATATCTATATTATCTTCAGAACGGGCGCGATATGTATGTGGACCAAGAGCTGGATATCAATCGGCTTTCGGATTATGACGTCGATGCAATCGTTCCGCAAAGTTTCCTCAAGGATGACAGCATCGATAACAAAGTGTTAACCCGTAGCGATAAAAACCGTGGCAAAAGTGATAACGTGCCATCGGAAGAGGTAGTGAAAAAGATGAAAAATTATTGGCGCCAGCTGCTTAATGCTAAGCTGATTACGCAACGGAAATTTGATAATCTGACTAAAGCTGAGCGGGGCGGGCTGTCGGAACTCGATAAAGCTGGGTTCATAAAAAGACAATTGGTCGAGACGCGCCAGATTACCAAACACGTAGCCCAAATTCTGGATTCACGGATGAATACGAAATACGACGAGAATGATAAACTGATCCGGGAAGTGAAAGTGATCACTCTTAAATCTAAACTGGTATCAGACTTTAGAAAAGATTTCCAATTCTATAAGGTTCGTGAAATCAACAATTATCATCACGCTCATGACGCCTATTTAAATGCGGTGGTAGGCACAGCGTTAATCAAGAAATATCCGAAACTTGAGTCAGAGTTTGTGTATGGAGATTATAAAGTTTATGATGTTAGAAAAATGATTGCTAAATCGGAGCAAGAAATTGGTAAGGCTACAGCGAAATATTTTTTTTACTCCAATATTATGAATTTCTTTAAGACTGAAATTACCTTGGCTAATGGTGAAATCCGCAAACGGCCGCTGATCGAAACAAATGGTGAAACTGGCGAGATTGTTTGGGATAAGGGCCGCGATTTCGCAACGGTCCGTAAGGTGTTGAGTATGCCTCAGGTGAATATTGTAAAGAAAACGGAAGTACAGACTGGCGGGTTTAGCAAAGAATCAATTTTGCCGAAACGTAATTCTGATAAACTGATCGCTCGTAAGAAAGATTGGGATCCCAAGAAGTATGGTGGCTTTGATAGTCCGACCGTAGCGTATAGTGTTCTGGTGGTAGCTAAGGTAGAAAAAGGCAAATCAAAGAAACTGAAGTCCGTGAAGGAACTACTGGGTATCACGATCATGGAACGATCAAGTTTTGAAAAAAATCCGATCGACTTCCTTGAAGCGAAAGGGTACAAAGAGGTAAAGAAGGACCTAATTATAAAACTGCCCAAATATAGTCTGTTTGAACTGGAAAATGGAAGGAAACGAATGTTGGCAAGCGCTGGTGAGCTGCAGAAAGGCAATGAATTAGCCCTGCCCTCGAAATATGTAAATTTCTTATACTTGGCCAGTCATTATGAAAAGCTGAAAGGATCACCTGAAGATAATGAACAAAAACAGCTCTTCGTCGAGCAACACAAGCATTACTTAGATGAAATTATAGAACAAATTTCTGAATTCTCAAAACGTGTTATATTAGCGGACGCGAATCTGGATAAGGTCTTGAGCGCATATAACAAGCATCGCGATAAACCGATACGGGAACAAGCGGAAAATATTATACATCTGTTCACGCTGACCAATTTAGGCGCGCCTGCTGCGTTTAAATATTTTGATACAACCATAGATCGTAAACGTTACACCAGCACAAAGGAAGTCCTGGATGCCACACTGATCCATCAGAGCATAACAGGACTGTACGAGACCCGTATCGATCTGAGTCAGCTTGGAGGTGATGGTGGTTCGGATTACAAGGATGACGATGATAAAGGTGGAAGTCCGAAAAAAAAAAGGAAAGTTTGATAACTCGAGTAAGGATCAGTAAGGATCTCCAGGCATC']))\n",
    "\n",
    "#findHomology('TCCCTATCAGTGATAGAGAAAAGAGAATCGGATCCAAAAGATCTAAAGAGGAGAAAGGATCTATGTTAATTGGCTATGTCCGTGTGTCCACCAACGATCAAAACACAGATCTTCAAAGAAATGCCCTCGTTTGCGCGGGATGTGAGCAGATCTTTGAAGATAAACTTAGCGGTACCCGTACCGATCGACCGGGTCTTAAACGCGCACTTAAACGGCTTCAGAAGGGGGATACCTTGGTCGTATGGAAACTGGACCGGCTGGGTAGATCGATGAAACATCTGATTAGTTTAGTTGGAGAACTGCGAGAACGCGGCATCAATTTTCGCTCCTTAACCGATAGCATCGATACCTCCAGTCCAATGGGGAGGTTTTTCTTCTATGTTATGGGCGCGCTGGCCGAAATGGAACGTGAATTGATTATCGAGCGGACAATGGCTGGGTTAGCGGCAGCACGAAATAAAGGACGGCGCTTCGGTCGTCCACCCAAGGGCGGCTCCGGTGGCTCGGGAGGTTCCGGCGGTAGCGGGGGCAGCGGCGGCAGTGGGGGCTCAGGAGGTTCTGATAAAAAATATTCTATAGGACTGGCTATTGGTACCAATTCAGTGGGTTGGGCGGTAATTACCGACGAATATAAAGTGCCGTCTAAAAAATTTAAGGTCTTGGGGAATACAGACCGTCACAGTATAAAAAAAAATCTGATCGGTGCGCTGCTGTTTGACTCTGGAGAAACTGCCGAAGCAACTCGTTTAAAACGAACCGCACGCCGTCGCTATACAAGGCGGAAAAATCGTATATGCTATTTACAAGAAATTTTTTCCAATGAAATGGCAAAAGTCGATGACAGCTTCTTTCATCGCTTAGAAGAGTCCTTTTTAGTTGAAGAAGATAAAAAACACGAACGCCATCCAATTTTTGGGAATATAGTTGATGAAGTCGCCTATCACGAAAAATATCCGACTATTTACCACCTGCGTAAGAAACTTGTGGATTCGACGGATAAAGCAGATCTGAGGCTGATTTATCTCGCCCTCGCACATATGATTAAATTCAGAGGTCATTTCCTGATAGAAGGCGATCTGAATCCCGACAACTCTGATGTTGATAAGTTGTTTATACAGCTGGTGCAAACCTATAATCAACTATTCGAAGAGAATCCTATAAACGCCAGCGGAGTGGACGCTAAAGCGATACTCAGCGCGCGCTTATCTAAAAGCCGGCGGCTGGAAAATCTGATCGCACAATTGCCAGGTGAAAAAAAAAATGGTTTATTTGGCAATCTGATAGCGCTTTCACTGGGGCTAACACCCAATTTTAAATCGAATTTTGATTTAGCCGAAGATGCTAAGCTGCAGCTGAGTAAAGATACCTATGATGACGATCTTGATAACCTGCTGGCACAAATTGGAGATCAATACGCAGATCTGTTTTTGGCCGCCAAGAATCTGTCGGATGCAATTCTGCTGAGCGATATTTTGAGGGTGAATACTGAAATCACTAAGGCACCGCTCTCTGCAAGTATGATCAAGCGTTATGATGAGCATCACCAGGACTTAACCCTGTTAAAAGCTCTGGTCCGCCAGCAGCTCCCGGAAAAATACAAGGAAATATTCTTCGATCAATCTAAGAATGGGTATGCCGGCTATATCGATGGCGGAGCGTCTCAGGAAGAATTTTATAAATTCATAAAACCTATCCTTGAAAAAATGGATGGAACGGAAGAGCTGCTGGTTAAACTGAATAGAGAGGATCTGTTACGCAAGCAGAGGACATTCGATAACGGTAGTATCCCACATCAAATACATCTGGGTGAACTTCACGCGATTCTTCGCCGCCAGGAAGACTTTTATCCCTTTCTAAAAGACAACAGGGAGAAAATTGAAAAAATCCTGACGTTTCGTATTCCTTATTATGTGGGGCCGCTGGCTCGTGGAAACAGTCGGTTCGCATGGATGACGCGCAAGTCCGAAGAGACGATAACGCCGTGGAATTTCGAAGAAGTGGTCGACAAAGGCGCAAGTGCACAGTCATTCATCGAACGTATGACTAACTTCGATAAAAATTTACCAAACGAGAAAGTTCTGCCTAAGCATAGCCTTCTCTATGAATACTTTACAGTCTATAACGAGTTAACTAAAGTTAAATATGTGACTGAAGGTATGCGTAAACCGGCGTTTCTGAGCGGCGAACAAAAAAAGGCAATTGTCGATCTGTTATTTAAAACAAATCGTAAAGTTACGGTCAAGCAATTGAAAGAGGACTACTTCAAAAAAATTGAGTGTTTCGACAGCGTCGAAATTTCTGGCGTTGAAGACCGTTTCAAT', 'CGTTGAAGACCGTTTCAATGCGTCGCTGGGAACTTACCACGATCTACTCAAAATCATAAAAGATAAAGACTTTCTGGACAATGAAGAGAACGAAGACATACTTGAAGACATCGTTCTGACCCTTACGCTATTTGAAGATAGGGAGATGATCGAAGAGCGCTTGAAGACCTATGCGCACCTATTCGATGATAAAGTTATGAAACAGCTCAAGCGGCGCCGTTACACGGGGTGGGGGCGTCTGTCGAGGAAGTTAATCAATGGAATTCGTGATAAGCAGTCAGGCAAGACGATCCTGGACTTTCTGAAATCAGATGGGTTTGCAAATCGTAACTTCATGCAGCTGATCCATGATGATTCGTTAACTTTTAAAGAAGATATTCAAAAAGCCCAAGTATCTGGTCAGGGTGACTCTCTTCACGAACATATCGCGAATTTGGCAGGTTCACCAGCTATTAAAAAAGGTATCTTGCAGACTGTGAAAGTTGTCGATGAACTTGTTAAAGTTATGGGTCGTCACAAGCCAGAGAATATCGTAATTGAGATGGCGCGTGAAAATCAAACAACCCAAAAAGGCCAGAAAAACTCGAGGGAACGCATGAAACGTATTGAAGAAGGTATTAAGGAGTTGGGTTCCCAGATACTTAAAGAACACCCGGTCGAAAATACGCAACTGCAAAATGAAAAATTATATCTATATTATCTTCAGAACGGGCGCGATATGTATGTGGACCAAGAGCTGGATATCAATCGGCTTTCGGATTATGACGTCGATGCAATCGTTCCGCAAAGTTTCCTCAAGGATGACAGCATCGATAACAAAGTGTTAACCCGTAGCGATAAAAACCGTGGCAAAAGTGATAACGTGCCATCGGAAGAGGTAGTGAAAAAGATGAAAAATTATTGGCGCCAGCTGCTTAATGCTAAGCTGATTACGCAACGGAAATTTGATAATCTGACTAAAGCTGAGCGGGGCGGGCTGTCGGAACTCGATAAAGCTGGGTTCATAAAAAGACAATTGGTCGAGACGCGCCAGATTACCAAACACGTAGCCCAAATTCTGGATTCACGGATGAATACGAAATACGACGAGAATGATAAACTGATCCGGGAAGTGAAAGTGATCACTCTTAAATCTAAACTGGTATCAGACTTTAGAAAAGATTTCCAATTCTATAAGGTTCGTGAAATCAACAATTATCATCACGCTCATGACGCCTATTTAAATGCGGTGGTAGGCACAGCGTTAATCAAGAAATATCCGAAACTTGAGTCAGAGTTTGTGTATGGAGATTATAAAGTTTATGATGTTAGAAAAATGATTGCTAAATCGGAGCAAGAAATTGGTAAGGCTACAGCGAAATATTTTTTTTACTCCAATATTATGAATTTCTTTAAGACTGAAATTACCTTGGCTAATGGTGAAATCCGCAAACGGCCGCTGATCGAAACAAATGGTGAAACTGGCGAGATTGTTTGGGATAAGGGCCGCGATTTCGCAACGGTCCGTAAGGTGTTGAGTATGCCTCAGGTGAATATTGTAAAGAAAACGGAAGTACAGACTGGCGGGTTTAGCAAAGAATCAATTTTGCCGAAACGTAATTCTGATAAACTGATCGCTCGTAAGAAAGATTGGGATCCCAAGAAGTATGGTGGCTTTGATAGTCCGACCGTAGCGTATAGTGTTCTGGTGGTAGCTAAGGTAGAAAAAGGCAAATCAAAGAAACTGAAGTCCGTGAAGGAACTACTGGGTATCACGATCATGGAACGATCAAGTTTTGAAAAAAATCCGATCGACTTCCTTGAAGCGAAAGGGTACAAAGAGGTAAAGAAGGACCTAATTATAAAACTGCCCAAATATAGTCTGTTTGAACTGGAAAATGGAAGGAAACGAATGTTGGCAAGCGCTGGTGAGCTGCAGAAAGGCAATGAATTAGCCCTGCCCTCGAAATATGTAAATTTCTTATACTTGGCCAGTCATTATGAAAAGCTGAAAGGATCACCTGAAGATAATGAACAAAAACAGCTCTTCGTCGAGCAACACAAGCATTACTTAGATGAAATTATAGAACAAATTTCTGAATTCTCAAAACGTGTTATATTAGCGGACGCGAATCTGGATAAGGTCTTGAGCGCATATAACAAGCATCGCGATAAACCGATACGGGAACAAGCGGAAAATATTATACATCTGTTCACGCTGACCAATTTAGGCGCGCCTGCTGCGTTTAAATATTTTGATACAACCATAGATCGTAAACGTTACACCAGCACAAAGGAAGTCCTGGATGCCACACTGATCCATCAGAGCATAACAGGACTGTACGAGACCCGTATCGATCTGAGTCAGCTTGGAGGTGATGGTGGTTCGGATTACAAGGATGACGATGATAAAGGTGGAAGTCCGAAAAAAAAAAGGAAAGTTTGATAACTCGAGTAAGGATCAGTAAGGATCTCCAGGCATC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import primer3\n",
    "import pandas\n",
    "import math\n",
    "sequences = []\n",
    "breakdown = []\n",
    "data = pandas.read_excel(\"/Users/aleclourenco/Desktop/REU-2017-Endy-Lab/PURE Datasheet.xlsx\", sheetname=\"List of PURE Components\", skiprows=[0])\n",
    "\n",
    "class DNASequence():\n",
    "    def __init__(self, sequence):\n",
    "        #Initialize the DNA variable, which has sequence info as well as additional properties (Tm, homodimer & hairpin formation)\n",
    "        self.sequence = sequence\n",
    "        self.composition = self.getComposition(sequence)\n",
    "        if len(sequence) != 0:\n",
    "            self.gcContent = self.getGCcontent(sequence)\n",
    "        else:\n",
    "            self.gcContent = 0\n",
    "        if len(sequence) < 60:\n",
    "            self.Tm = primer3.calcTm(sequence)\n",
    "            self.Homodimer = primer3.calcHomodimer(sequence)\n",
    "            self.Hairpin = primer3.calcHairpin(sequence)\n",
    "        else:\n",
    "            self.Tm = 100000 #set the Tm arbitrarily high\n",
    "        \n",
    "    def getComposition(self, sequence):\n",
    "        composition = {\"A\" : 0, \"G\" : 0, \"C\" : 0, \"T\" : 0}\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                composition[\"A\"] = composition[\"A\"] + 1\n",
    "            elif base == \"T\":\n",
    "                composition[\"T\"] = composition[\"T\"] + 1\n",
    "            elif base == \"G\":\n",
    "                composition[\"G\"] = composition[\"G\"] + 1\n",
    "            elif base == \"C\":\n",
    "                composition[\"C\"] = composition[\"C\"] + 1\n",
    "            else:\n",
    "                print(\"error 2\")\n",
    "        return composition\n",
    "    def getGCcontent(self, sequence):\n",
    "        GCpercent = (self.composition[\"G\"] + self.composition[\"C\"])/(self.composition[\"G\"] + self.composition[\"T\"] + self.composition[\"A\"] + self.composition[\"C\"])\n",
    "        return GCpercent\n",
    "    def reverseComplement(self, sequence):\n",
    "        #Takes the complement of the sequence and reverses it.\n",
    "        reverseComplement = \"\"\n",
    "        for base in sequence:\n",
    "            if base == \"A\":\n",
    "                reverseComplement = \"T\" + reverseComplement\n",
    "            elif base == \"T\":\n",
    "                reverseComplement = \"A\" + reverseComplement\n",
    "            elif base == \"G\":\n",
    "                reverseComplement = \"C\" + reverseComplement\n",
    "            elif base == \"C\":\n",
    "                reverseComplement = \"G\" + reverseComplement \n",
    "            else:\n",
    "                print (\"error\")\n",
    "        return (reverseComplement)\n",
    "def reverseComplement(sequence):\n",
    "    #Takes the complement of the sequence and reverses it.\n",
    "    reverseComplement = \"\"\n",
    "    for base in sequence:\n",
    "        if base == \"A\":\n",
    "            reverseComplement = \"T\" + reverseComplement \n",
    "        elif base == \"T\":\n",
    "            reverseComplement = \"A\" + reverseComplement\n",
    "        elif base == \"G\":\n",
    "            reverseComplement = \"C\" + reverseComplement\n",
    "        elif base == \"C\":\n",
    "            reverseComplement = \"G\" + reverseComplement\n",
    "        else:\n",
    "            print (\"error\")\n",
    "    return (reverseComplement)\n",
    "def screenPrimers(insert, upstreamBackbone, downstreamBackbone):\n",
    "    #run the main program to create and screen the forward and reverse primers and then add overhangs.\n",
    "    forwardPrimer = screenPrimer(insert, \"F\")\n",
    "    #print(forwardPrimer.sequence, \"forward\")\n",
    "    reversePrimer = screenPrimer(insert, \"R\")\n",
    "    #print(reversePrimer.sequence, \"reverse\")\n",
    "    if len(forwardPrimer.sequence) and len(reversePrimer.sequence) != 0:\n",
    "        primers = addOverhangs(forwardPrimer, reversePrimer, \"reverse\", upstreamBackbone, downstreamBackbone)\n",
    "        return (primers[0].sequence, primers[1].sequence)\n",
    "    else:\n",
    "        return (\"You cannot make primers with the given parameters\", \"You cannot make primers with the given parameters\")\n",
    "    \n",
    "def screenPrimer(insert, whichPrimer, primerSize=18): #Make primers for an insert to amplify it such that it can be assembled with Gibson\n",
    "    if primerSize < 40:\n",
    "        #Make sure you are selecting the right region to prime. \n",
    "        #This algorithm creates primers that amplify the WHOLE SEQUENCE minus the stop codon at the end.\n",
    "        if whichPrimer == \"F\": \n",
    "            primer = DNASequence(insert.sequence[:primerSize])\n",
    "        elif whichPrimer == \"R\":\n",
    "            primer = DNASequence(reverseComplement(insert.sequence[len(insert.sequence) - primerSize: len(insert.sequence)]))\n",
    "        else:\n",
    "            print(\"error\")\n",
    "        #Add various screens below here. You can update the parameters to be more or less stringent depending on how likely you would like your primers to work\n",
    "        isGood = True\n",
    "        if primer.Tm < 58.0 or primer.Tm > 59.5:\n",
    "            isGood = False\n",
    "        if float(primer.Homodimer.dg) < -10000.0:\n",
    "            isGood = False\n",
    "        if float(primer.Hairpin.dg) < -5000.0:\n",
    "            isGood = False\n",
    "    \n",
    "        #return the shortest (18bp â‰¤) desireable primers\n",
    "        if isGood == True:\n",
    "            #screening returned a good primer; return the primer\n",
    "            return primer\n",
    "        else:\n",
    "            #primer failed screening, try again with a larger primer using recursion\n",
    "            if whichPrimer == \"F\":\n",
    "                primer = screenPrimer(insert, \"F\", primerSize + 1) \n",
    "            elif whichPrimer == \"R\":\n",
    "                primer = screenPrimer(insert, \"R\", primerSize + 1)\n",
    "            return primer\n",
    "    else:\n",
    "        #parameters are too stringent; no primers were found that matched the criteria given. Return an empty string\n",
    "        return DNASequence(\"\")\n",
    "def addOverhangs(primer1, primer2, orientation, leftBackbone, rightBackbone):\n",
    "    #decide how the construct should be oriented in the final vector and add the overhangs so Gibson will add correctly.\n",
    "    if orientation == \"forward\":\n",
    "        primer2.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer2.sequence\n",
    "        primer1.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer1.sequence\n",
    "        return primer1, primer2\n",
    "    elif orientation == \"reverse\":\n",
    "        primer1.sequence = reverseComplement(rightbackbone.sequence[:40]) + primer1.sequence\n",
    "        primer2.sequence = leftBackbone.sequence[len(leftBackbone.sequence) - 40: len(leftBackbone.sequence)] + primer2.sequence\n",
    "        return primer1, primer2\n",
    "    else:\n",
    "        print (\"please enter either a forward or reverse strand orientation\")\n",
    "def formatSeq(oldSeq):\n",
    "    if not type(oldSeq) is str:\n",
    "        oldSeq = \"\"\n",
    "    new = oldSeq.replace(\"\\n\", \"\")\n",
    "    for number in range(0,10):\n",
    "        temp = new.replace(str(number), \"\")\n",
    "        new = temp\n",
    "    oldSeq = new\n",
    "    new = oldSeq.replace(\" \", \"\")\n",
    "    return new.upper()\n",
    "\n",
    "def appendPrimers(data, insert, tag, primerList, nameList):\n",
    "    name = str(data.ix[number,\"Protein Name\"]).replace(\" \", \"\") \n",
    "    nameFOR = name + \" FOR \" + tag + \"-tagged\"\n",
    "    nameREV = name + \" REV \" + tag + \"-tagged\"\n",
    "    primerList.append(screenPrimers(insert, leftbackbone , rightbackbone)[0])\n",
    "    nameList.append(nameFOR)\n",
    "    primerList.append(screenPrimers(insert, leftbackbone , rightbackbone)[1])\n",
    "    nameList.append(nameREV)\n",
    "\n",
    "\n",
    "primerList = []\n",
    "nameList = []\n",
    "\n",
    "for number in range(0,49):\n",
    "    tag = \"\"\n",
    "    if (data.ix[number, \"His Tagged Where?\"]) == \"C\":\n",
    "        tag = \"C\"\n",
    "        #Correct the backbone so that the primers generated add in the correct area of the backbone\n",
    "        leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg'.upper())\n",
    "        rightbackbone = DNASequence('ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcgcgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgcca'.upper())\n",
    "        insert = DNASequence(formatSeq(data.ix[number,\"Nucleotide Sequence\"]))\n",
    "        if len(insert.sequence) != 0:\n",
    "            appendPrimers(data, insert, tag, primerList, nameList)\n",
    "    elif (data.ix[number, \"His Tagged Where?\"]) == \"N\": \n",
    "        tag = \"N\"\n",
    "        #Correct the backbone so that the primers generated add in the correct area of the backbone\n",
    "        leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatc'.upper())\n",
    "        rightbackbone = DNASequence('atggctgccgcgcggcaccaggccgctgctgtgatgatgatgatgatggctgctgcccatggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccag'.upper())\n",
    "        insert = DNASequence(formatSeq(data.ix[number,\"Nucleotide Sequence\"]))\n",
    "        if len(insert.sequence) != 0:\n",
    "            appendPrimers(data, insert, tag, primerList, nameList)\n",
    "    else: #We don't know whether unspecified proteins are N or C His-tagged, so we assume they could be either.\n",
    "        leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatctcagtggtggtggtggtggtg'.upper())\n",
    "        rightbackbone = DNASequence('ggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccagccacgtttctgcgaaaacgcgggaaaaagtggaagcggcgatggcggagctgaattacattcccaaccgcgtggcacaacaactggcgggcaaacagtcgttgctgattggcgttgccacctccagtctggccctgcacgcgccgtcgcaaattgtcgcggcgattaaatctcgcgccgatcaactgggtgccagcgtggtggtgtcgatggtagaacgaagcggcgtcgaagcctgtaaagcggcggtgcacaatcttctcgcgcaacgcgtcagtgggctgatcattaactatccgctggatgaccaggatgccattgctgtggaagctgcctgcactaatgttccggcgttatttcttgatgtctctgaccagacacccatcaacagtattattttctcccatgaagacggtacgcgactgggcgtggagcatctggtcgcattgggtcaccagcaaatcgcgctgttagcgggcccattaagttctgtctcggcgcgtctgcgtctggctggctggcataaatatctcactcgcaatcaaattcagccgatagcggaacgggaaggcgactggagtgccatgtccggttttcaacaaaccatgcaaatgctgaatgagggcatcgttcccactgcgatgctggttgccaacgatcagatggcgctgggcgcaatgcgcgccattaccgagtccgggctgcgcgttggtgcggatatctcggtagtgggatacgacgataccgaagacagctcatgttatatcccgccgttaaccaccatcaaacaggattttcgcctgctggggcaaaccagcgtggaccgcttgctgcaactctctcagggccaggcggtgaagggcaatcagctgttgcccgtctcactggtgaaaagaaaaaccaccctggcgcccaatacgcaaaccgcctctccccgcgcgttggccgattcattaatgcagctggcacgacaggtttcccgactggaaagcgggcagtgagcgcaacgcaattaatgtaagttagctcactcattaggcaccgggatctcgaccgatgcccttgagagccttcaacccagtcagctccttccggtgggcgcggggcatgactatcgtcgccgcacttatgactgtcttctttatcatgcaactcgtaggacaggtgccggcagcgctctgggtcattttcggcgaggaccgctttcgctggagcgcgacgatgatcggcctgtcgcttgcggtattcggaatcttgcacgccctcgctcaagccttcgtcactggtcccgccaccaaacgtttcggcgagaagcaggccattatcgccggcatggcggccccacgggtgcgcatgatcgtgctcctgtcgttgaggacccggctaggctggcggggttgccttactggttagcagaatgaatcaccgatacgcgagcgaacgtgaagcgactgctgctgcaaaacgtctgcgacctgagcaacaacatgaatggtcttcggtttccgtgtttcgtaaagtctggaaacgcggaagtcagcgccctgcaccattatgttccggatctgcatcgcaggatgctgctggctaccctgtggaacacctacatctgtattaacgaagcgctggcattgaccctgagtgatttttctctggtcccgccgcatccataccgccagttgtttaccctcacaacgttccagtaaccgggcatgttcatcatcagtaacccgtatcgtgagcatcctctctcgtttcatcggtatcattacccccatgaacagaaatcccccttacacggaggcatcagtgaccaaacaggaaaaaaccgcccttaacatggcccgctttatcagaagccagacattaacgcttctggagaaactcaacgagctggacgcggatgaacaggcagacatctgtgaatcgcttcacgaccacgctgatgagctttaccgcagctgcctcgcgcgtttcggtgatgacggtgaaaacctctgacacatgcagctcccggagacggtcacagcttgtctgtaagcggatgccgggagcagacaagcccgtcagggcgcgtcagcgggtgttggcgggtgtcggggcgcagccatgacccagtcacgtagcgatagcggagtgtatactggcttaactatgcggcatcagagcagattgtactgagagtgcaccatatatgcggtgtgaaataccgcacagatgcgtaaggagaaaataccgcatcaggcgctcttccgcttcctcgctcactgactcgctgcgctcggtcgttcggctgcggcgagcggtatcagctcactcaaaggcggtaatacggttatccacagaatcaggggataacgcaggaaagaacatgtgagcaaaaggccagcaaaaggccaggaaccgtaaaaaggccgcgttgctggcgtttttccataggctccgcccccctgacgagcatcacaaaaatcgacgctcaagtcagaggtggcgaaacccgacaggactataaagataccaggcgtttccccctggaagctccctcgtgcgctctcctgttccgaccctgccgcttaccggatacctgtccgcctttctcccttcgggaagcgtggcgctttctcatagctcacgctgtaggtatctcagttcggtgtaggtcgttcgctccaagctgggctgtgtgcacgaaccccccgttcagcccgaccgctgcgccttatccggtaactatcgtcttgagtccaacccggtaagacacgacttatcgccactggcagcagccactggtaacaggattagcagagcgaggtatgtaggcggtgctacagagttcttgaagtggtggcctaactacggctacactagaaggacagtatttggtatctgcgctctgctgaagccagttaccttcggaaaaagagttggtagctcttgatccggcaaacaaaccaccgctggtagcggtggtttttttgtttgcaagcagcagattacgcgcagaaaaaaaggatctcaagaagatcctttgatcttttctacggggtctgacgctcagtggaacgaaaactcacgttaagggattttggtcatgaacaataaaactgtctgcttacataaacagtaatacaaggggtgttatgagccatattcaacgggaaacgtcttgctctaggccgcgattaaattccaacatggatgctgatttatatgggtataaatgggctcgcgataatgtcgggcaatcaggtgcgacaatctatcgattgtatgggaagcccgatgcgccagagttgtttctgaaacatggcaaaggtagcgttgccaatgatgttacagatgagatggtcagactaaactggctgacggaatttatgcctcttccgaccatcaagcattttatccgtactcctgatgatgcatggttactcaccactgcgatccccgggaaaacagcattccaggtattagaagaatatcctgattcaggtgaaaatattgttgatgcgctggcagtgttcctgcgccggttgcattcgattcctgtttgtaattgtccttttaacagcgatcgcgtatttcgtctcgctcaggcgcaatcacgaatgaataacggtttggttgatgcgagtgattttgatgacgagcgtaatggctggcctgttgaacaagtctggaaagaaatgcataaacttttgccattctcaccggattcagtcgtcactcatggtgatttctcacttgataaccttatttttgacgaggggaaattaataggttgtattgatgttggacgagtcggaatcgcagaccgataccaggatcttgccatcctatggaactgcctcggtgagttttctccttcattacagaaacggctttttcaaaaatatggtattgataatcctgatatgaataaattgcagtttcatttgatgctcgatgagtttttctaagaattaattcatgagcggatacatatttgaatgtatttagaaaaataaacaaataggggttccgcgcacatttccccgaaaagtgccacctgaaattgtaaacgttaatattttgttaaaattcgcgttaaatttttgttaaatcagctcattttttaaccaataggccgaaatcggcaaaatcccttataaatcaaaagaatagaccgagatagggttgagtgttgttccagtttggaacaagagtccactattaaagaacgtggactccaacgtcaaagggcgaaaaaccgtctatcagggcgatggcccactacgtgaaccatcaccctaatcaagttttttggggtcgaggtgccgtaaagcactaaatcggaaccctaaagggagcccccgatttagagcttgacggggaaagccggcgaacgtggcgagaaaggaagggaagaaagcgaaaggagcgggcgctagggcgctggcaagtgtagcggtcacgctgcgcgtaaccaccacacccgccgcgcttaatgcgccgctacagggcgcgtcccattcgcca'.upper())\n",
    "        insert = DNASequence(formatSeq(data.ix[number,\"Nucleotide Sequence\"]))\n",
    "        if len(insert.sequence) != 0:\n",
    "            tag = \"C\"\n",
    "            appendPrimers(data, insert, tag, primerList, nameList)\n",
    "        leftbackbone = DNASequence('atccggatatagttcctcctttcagcaaaaaacccctcaagacccgtttagaggccccaaggggttatgctagttattgctcagcggtggcagcagccaactcagcttcctttcgggctttgttagcagccggatc'.upper())\n",
    "        rightbackbone = DNASequence('atggctgccgcgcggcaccaggccgctgctgtgatgatgatgatgatggctgctgcccatggtatatctccttcttaaagttaaacaaaattatttctagaggggaattgttatccgctcacaattcccctatagtgagtcgtattaatttcgcgggatcgagatctcgatcctctacgccggacgcatcgtggccggcatcaccggcgccacaggtgcggttgctggcgcctatatcgccgacatcaccgatggggaagatcgggctcgccacttcgggctcatgagcgcttgtttcggcgtgggtatggtggcaggccccgtggccgggggactgttgggcgccatctccttgcatgcaccattccttgcggcggcggtgctcaacggcctcaacctactactgggctgcttcctaatgcaggagtcgcataagggagagcgtcgagatcccggacaccatcgaatggcgcaaaacctttcgcggtatggcatgatagcgcccggaagagagtcaattcagggtggtgaatgtgaaaccagtaacgttatacgatgtcgcagagtatgccggtgtctcttatcagaccgtttcccgcgtggtgaaccaggccag'.upper())\n",
    "        if len(insert.sequence) != 0:\n",
    "            tag = \"N\"\n",
    "            appendPrimers(data, insert, tag, primerList, nameList)\n",
    "        \n",
    "dataframe = pandas.DataFrame(primerList, nameList)\n",
    "\n",
    "writer = pandas.ExcelWriter('test2.xlsx', engine='openpyxl')\n",
    "\n",
    "dataframe.to_excel(writer)\n",
    "\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'atgGCCAAAG AAGACAATAT TGAAATGCAA GGTACCGTTC TTGAAACGTT GCCTAATACC\\nATGTTCCGCG TAGAGTTAGA AAACGGTCAC GTGGTTACTG CACACATCTC CGGTAAAATG\\nCGCAAAAACT ACATCCGCAT CCTGACGGGC GACAAAGTGA CTGTTGAACT GACCCCGTAC\\nGACCTGAGCA AAGGCCGCAT TGTCTTCCGT AGTCGCtga'"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.ix[1,'Nucleotide Sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Abbreviation', 'Gene Name', 'Protein Name', 'Function', 'Reference',\n",
       "       'Sequence', 'Nucleotide Sequence', 'Amount Added (ug)', 'AA Sequence',\n",
       "       'Methionine', 'Alanine', 'Arginine', 'Asparagine', 'Aspartic Acid',\n",
       "       'Cysteine', 'Glutamic Acid', 'Glutamine', 'Glycine', 'Histidine',\n",
       "       'Isoleucine', 'Leucine', 'Lysine', 'Phenylalanine', 'Proline', 'Serine',\n",
       "       'Threonine', 'Tryptophan', 'Tyrosine', 'Valine'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abbreviation</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Protein Name</th>\n",
       "      <th>Function</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Nucleotide Sequence</th>\n",
       "      <th>Amount Added (ug)</th>\n",
       "      <th>AA Sequence</th>\n",
       "      <th>Methionine</th>\n",
       "      <th>...</th>\n",
       "      <th>Isoleucine</th>\n",
       "      <th>Leucine</th>\n",
       "      <th>Lysine</th>\n",
       "      <th>Phenylalanine</th>\n",
       "      <th>Proline</th>\n",
       "      <th>Serine</th>\n",
       "      <th>Threonine</th>\n",
       "      <th>Tryptophan</th>\n",
       "      <th>Tyrosine</th>\n",
       "      <th>Valine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Transcription/Translation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IF1</td>\n",
       "      <td>infA</td>\n",
       "      <td>Initiation Factor 1</td>\n",
       "      <td>Not fully understood but it is suggested that ...</td>\n",
       "      <td>Grunberg-Manago et al. 2014</td>\n",
       "      <td>NP_415404</td>\n",
       "      <td>atgGCCAAAG AAGACAATAT TGAAATGCAA GGTACCGTTC TT...</td>\n",
       "      <td>1</td>\n",
       "      <td>MAKEDNIEMQ GTVLETLPNT MFRVELENGH VVTAHISGKM RK...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IF2</td>\n",
       "      <td>infB</td>\n",
       "      <td>Initiation Factor 2</td>\n",
       "      <td>IF2 is implicated in fMet-tRNA_fMet binding to...</td>\n",
       "      <td>Grunberg-Manago et al. 2014</td>\n",
       "      <td>NP_417637</td>\n",
       "      <td>atgACAGATG TAACGATTAA AACGCTGGCC GCAGAGCGAC AG...</td>\n",
       "      <td>2</td>\n",
       "      <td>MTDVTIKTLA AERQTSVERL VQQFADAGIR KSADDSVSAQ EK...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IF3</td>\n",
       "      <td>infC</td>\n",
       "      <td>Initiation Factor 3</td>\n",
       "      <td>IF3 discriminates against noninitiator tRNAs b...</td>\n",
       "      <td>Grunberg-Manago et al. 2014</td>\n",
       "      <td>NP_416233</td>\n",
       "      <td>attAAAGGCG GAAAACGAGT TCAAACGGCG CGCCCTAACC GT...</td>\n",
       "      <td>0.75</td>\n",
       "      <td>MKGGKRVQTA RPNRINGEIR AQEVRLTGLE GEQLGIVSLR EA...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>12.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EF-G</td>\n",
       "      <td>fusA</td>\n",
       "      <td>Elongation Factor G</td>\n",
       "      <td>Facilitates the translocation of the ribosome ...</td>\n",
       "      <td>Chen et al. 2016</td>\n",
       "      <td>NP_417799</td>\n",
       "      <td>atgGCTCGTA CAACACCCAT CGCACGCTAC CGTAACATCG GT...</td>\n",
       "      <td>1</td>\n",
       "      <td>MARTTPIARY RNIGISAHID AGKTTTTERI LFYTGVNHKI GE...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>...</td>\n",
       "      <td>50.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Abbreviation Gene Name         Protein Name  \\\n",
       "0  Transcription/Translation       NaN                  NaN   \n",
       "1                        IF1      infA  Initiation Factor 1   \n",
       "2                        IF2      infB  Initiation Factor 2   \n",
       "3                        IF3      infC  Initiation Factor 3   \n",
       "4                       EF-G      fusA  Elongation Factor G   \n",
       "\n",
       "                                            Function  \\\n",
       "0                                                NaN   \n",
       "1  Not fully understood but it is suggested that ...   \n",
       "2  IF2 is implicated in fMet-tRNA_fMet binding to...   \n",
       "3  IF3 discriminates against noninitiator tRNAs b...   \n",
       "4  Facilitates the translocation of the ribosome ...   \n",
       "\n",
       "                     Reference   Sequence  \\\n",
       "0                          NaN        NaN   \n",
       "1  Grunberg-Manago et al. 2014  NP_415404   \n",
       "2  Grunberg-Manago et al. 2014  NP_417637   \n",
       "3  Grunberg-Manago et al. 2014  NP_416233   \n",
       "4             Chen et al. 2016  NP_417799   \n",
       "\n",
       "                                 Nucleotide Sequence Amount Added (ug)  \\\n",
       "0                                                NaN               NaN   \n",
       "1  atgGCCAAAG AAGACAATAT TGAAATGCAA GGTACCGTTC TT...                 1   \n",
       "2  atgACAGATG TAACGATTAA AACGCTGGCC GCAGAGCGAC AG...                 2   \n",
       "3  attAAAGGCG GAAAACGAGT TCAAACGGCG CGCCCTAACC GT...              0.75   \n",
       "4  atgGCTCGTA CAACACCCAT CGCACGCTAC CGTAACATCG GT...                 1   \n",
       "\n",
       "                                         AA Sequence  Methionine   ...    \\\n",
       "0                                                NaN         NaN   ...     \n",
       "1  MAKEDNIEMQ GTVLETLPNT MFRVELENGH VVTAHISGKM RK...         4.0   ...     \n",
       "2  MTDVTIKTLA AERQTSVERL VQQFADAGIR KSADDSVSAQ EK...        21.0   ...     \n",
       "3  MKGGKRVQTA RPNRINGEIR AQEVRLTGLE GEQLGIVSLR EA...         6.0   ...     \n",
       "4  MARTTPIARY RNIGISAHID AGKTTTTERI LFYTGVNHKI GE...        23.0   ...     \n",
       "\n",
       "   Isoleucine  Leucine  Lysine  Phenylalanine  Proline  Serine  Threonine  \\\n",
       "0         NaN      NaN     NaN            NaN      NaN     NaN        NaN   \n",
       "1         5.0      6.0     5.0            2.0      2.0     3.0        7.0   \n",
       "2        47.0     54.0    66.0           15.0     21.0    44.0       42.0   \n",
       "3        12.0     15.0    21.0            5.0      7.0     7.0        5.0   \n",
       "4        50.0     43.0    44.0           24.0     33.0    29.0       38.0   \n",
       "\n",
       "   Tryptophan  Tyrosine  Valine  \n",
       "0         NaN       NaN     NaN  \n",
       "1         0.0       2.0     7.0  \n",
       "2         2.0      11.0    83.0  \n",
       "3         0.0       3.0    14.0  \n",
       "4         6.0      19.0    60.0  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import datetime as dt\n",
    "import pandas as pd\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "def add_frame_to_workbook(filename, tabname, dataframe, timestamp):\n",
    "    \"\"\"\n",
    "    Save a dataframe to a workbook tab with the filename and tabname\n",
    "    coded to timestamp\n",
    "\n",
    "    :param filename: filename to create, can use strptime formatting\n",
    "    :param tabname: tabname to create, can use strptime formatting\n",
    "    :param dataframe: dataframe to save to workbook\n",
    "    :param timestamp: timestamp associated with dataframe\n",
    "    :return: None\n",
    "    \"\"\"\n",
    "    filename = timestamp.strftime(filename)\n",
    "    sheet_name = timestamp.strftime(tabname)\n",
    "\n",
    "    # create a writer for this month and year\n",
    "    writer = pd.ExcelWriter(filename, engine='openpyxl')\n",
    "\n",
    "    try:\n",
    "        # try to open an existing workbook\n",
    "        writer.book = load_workbook(filename)\n",
    "\n",
    "        # copy existing sheets\n",
    "        writer.sheets = dict(\n",
    "            (ws.title, ws) for ws in writer.book.worksheets)\n",
    "    except IOError:\n",
    "        # file does not exist yet, we will create it\n",
    "        pass\n",
    "\n",
    "    # write out the new sheet\n",
    "    dataframe.to_excel(writer, sheet_name=sheet_name)\n",
    "\n",
    "    # save the workbook\n",
    "    writer.save()\n",
    "\n",
    "data = [x.strip().split() for x in \"\"\"\n",
    "            Primer Name  Sequence\n",
    "                    FOR  ATGTGCCATGA...\n",
    "                    REV  TGCCTAGTCA...\n",
    "\"\"\".split('\\n')[1:-1]]\n",
    "df = pd.DataFrame(data=data[1:], columns=data[0])\n",
    "\n",
    "name_template = './sample-%m-%y.xlsx'\n",
    "tab_template = '%d_%H_%M'\n",
    "now = dt.datetime.now()\n",
    "in_an_hour = now + dt.timedelta(hours=1)\n",
    "add_frame_to_workbook(name_template, tab_template, df, now)\n",
    "add_frame_to_workbook(name_template, tab_template, df, in_an_hour)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Sequence      name\n",
      "0 -0.420077  0.081306\n",
      "1  0.578674  2.456459\n",
      "2 -0.355857  0.524907\n",
      "3 -0.450266       NaN\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "df = pd.DataFrame({'name' : pd.Series(np.random.randn(3)),'Sequence' : pd.Series(np.random.randn(4))})\n",
    "print(df)\n",
    "\n",
    "writer = pd.ExcelWriter('test2.xlsx', engine='openpyxl')\n",
    "\n",
    "df.to_excel(writer)\n",
    "\n",
    "writer.save()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
