---
layout: page
title: Biokernel
permalink: /biokernel/
weight: 5
---

The engineering of living matter grants accessibility and interaction with our natural world. The goal of synthetic biology is to make the engineering of biology easier [[1](https://www.ncbi.nlm.nih.gov/pubmed/16306983)]. Current practice of engineering biology is done largely in an ad hoc manner within complex preexisting organisms, whose cells and genomes having been complicated from billions of years of evolution optimized for robustness, environmental adaptation, lineage persistence, and without selection for human understanding. Advances in genetic engineering have provided an avenue for rapid and rational design choices to be made for improving the engineerability of organisms. These have led to substantial changes in genome sequences while maintaining or extending cell function, including by gene refactoring [[2](https://www.ncbi.nlm.nih.gov/pubmed/16729053)], full codon replacement [[3](https://www.ncbi.nlm.nih.gov/pubmed/19633652)], and full genome reconstruction [[4](https://www.ncbi.nlm.nih.gov/pubmed/20488990), [5](https://www.ncbi.nlm.nih.gov/pubmed/12114528), [6](https://www.ncbi.nlm.nih.gov/pubmed/26076868)]. However, all current genome design is templated using existing organisms. With exponential decrease in the cost of DNA synthesis and sequencing, de novo synthesis of entire genomes will certainly become tractable and preferable to piecewise modification. The bottleneck will then become our ability to design functional genomes encoding operational organisms.  

The prospect of reliably transferring and scaling human intentions into living matter necessitates the realization of a modellable, understandable cell. One strategy is to forward engineer an entire cell from scratch akin to origin of life conditions. Work in this paradigm includes chemically synthesizing the biomolecules necessary for cell function: lipid bilayer membranes, proteins, and transcription translation machinery, and mixing the constituents to spontaneously self-assemble into a synthetic protocell [[7](https://www.ncbi.nlm.nih.gov/pubmed/11201752), [8](https://www.ncbi.nlm.nih.gov/pubmed/24606140), [9](https://www.ncbi.nlm.nih.gov/pubmed/21941243), [10](https://www.ncbi.nlm.nih.gov/pubmed/15591347), [11](https://www.ncbi.nlm.nih.gov/pubmed/21317359), [12](https://www.ncbi.nlm.nih.gov/pubmed/21886681)]. Groups employing this bottom-up method have made progress on different aspects of cell synthesis, but the capacity to produce a complete replicating cell from first principles still remains a monumental challenge.


Another strategy for achieving an understandable cell is to reduce existing organisms into something capable of being understood. One approach has been to uncover the subset of genes, or minimal genome, essential for survival in a defined environment [[13](https://www.ncbi.nlm.nih.gov/pubmed/27013737), [14](https://www.ncbi.nlm.nih.gov/pubmed/16645050)]. Work to minimize the M. mycoides genome has produced a 531 kbp genome with 473 genes, smaller than the genome of any autonomously replicating organisms found in nature. Remarkably, 149 of those genes could not be specifically assigned a biological function and 55 of those have completely unknown function. Work is ongoing to elucidate the functions of these genes, but uncovering these may not produce a cell amenable to full-scale engineering. Natural genomes, even in a reduced form, are organized in a manner designed for robustness and adaptation across fluctuating environments. An architecture based upon fine-tuned performance and cost effective reuse of core functions, consistent with human designed non-biological systems [[15](https://www.ncbi.nlm.nih.gov/pubmed/21886681)], may provide a more powerful use case for current applications and expansion into new areas.


Both approaches have demonstrated significant progress and have provided invaluable insight into the core functional dependencies of self-replicating living matter, but have come short of a fully engineered or engineerable cell. The bottom-up forward engineering approach entails understanding through building up, but there remains much work to be done before the capacity to express and sustain genomes on the scale of natural organisms is realized. Though the reverse engineering top-down approach has succeeded in synthesizing and booting up an artificial chromosome inside of a host organism, the chromosome and its function were fundamentally derived from an extant organism and minimised by a rigid algorithm. We see this approach as valuable, but  If a synthetic cell is to be readily amenable to human intention, it must be composed of a defined set of functions whose constituents are understood and predetermined.


We wish to create an organism for which all molecules are understood from a first order functional perspective, enabling rational design of genomes not limited to single lineages. Our strategy for doing so is to use a combined approach that harnesses the strengths of the aforementioned methods.


We aim to bootstrap a cellular container—a cell-like object containing native cellular machinery but lacking native DNA—using a rationally designed, forward engineered genome to restore growth and division. This forward engineered genome -- a biokernel -- forms the core of a genetic operating system in which we understand the function and regulation of all genes. By dilution, the rebooted container grows to contain only that for which its genome encodes, leading to a system in which we can fully account for every molecule.


It is not our intention to recapitulate the native genome of any organism. Rather, we suspect that a forward engineered genetic operating system will look fundamentally different from those selected by evolution. Modularity, modelability, and simplicity of understanding may take precedence over concerns such as efficiency or redundancy. Drawing inspiration from operating systems such as Linux, we foresee a flatter functional call-stack, with rigorously characterised drivers of core functions and regulators of resources, wrapped in well-defined interfaces. The understandable system would serve as a platform for reliable implementation, the sharing and integration of new intracellular systems, as well as the branch point for a new diversity of novel organisms.


This task is too large for any one team. The current paradigm of groups of scientists and engineers working alone or within a small community has not realized an entire engineered cell. We believe a different approach is required, built on wide and deep networks of collaboration, and where complexity is controlled at the level at which it is created. We take inspiration from the open source communities which have powered the revolution in computer technology; a bazaar of creativity, facilitated by robust tools for communication, collaboration, and reuse, producing systems of incredible sophistication while empowering hobbyists and corporations alike to more easily realise their goals. Like Ritchie and Thompson building [UNIX](/engineering/about/unix-philosophy/), we hope that Build-A-Cell will provide a nexus “around which a community could form” to realise a goal as challenging as it is important.
